<html>
<head>
    <script>
        let base_api_url = "https://spliceailookup-api.broadinstitute.org"
        if (location.origin && location.origin.startsWith("http") && !location.origin.includes("broadinstitute.")) {
            base_api_url = location.origin
        }

        //base_api_url = "http://localhost:8080"

        console.log("Using SpliceAI-lookup API url:", base_api_url)
    </script>
    <meta content="width=device-width, initial-scale=1" charset="utf-8" name="viewport" />
    <title>SpliceAI Lookup</title>
    <link rel="shortcut icon" href="/icon.png">

    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.css" />
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.js"></script>
    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/underscore@1.12.0/underscore-min.js"></script>
    <script type="text/javascript" src="igv.min.js"></script>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-172722632-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());
      gtag('config', 'UA-172722632-1');
    </script>
    <style>
        :root {
          --CANONICAL_TRANSCRIPT_BACKGROUND_COLOR: #eaf3ff;
          --CODING_TRANSCRIPT_BACKGROUND_COLOR: white;
          --NON_CODING_TRANSCRIPT_BACKGROUND_COLOR: #e5e5e5;
        }

        .spliceai.coding {
          background-color: var(--CODING_TRANSCRIPT_BACKGROUND_COLOR);
        }

        .spliceai.noncoding {
          background-color: var(--NON_CODING_TRANSCRIPT_BACKGROUND_COLOR);
        }

        .spliceai.canonical {
          background-color: var(--CANONICAL_TRANSCRIPT_BACKGROUND_COLOR);
        }

        body {
            overflow-x: scroll;
        }

        body, .ui.form, label {
            font-size: 11pt !important;
        }

        td {
            padding: 5px 15px;
        }

        .upperBorder {
            border-top: 1px solid black;
        }

        #pangolin_header th {
          border-top: 1px solid rgba(34,36,38,.15);
        }

        table {
            font-size: 11pt !important;
            border-collapse: collapse;
            border-left: 0px !important;
            border-right: 0px !important;
            width: 100%;
        }

        th div {
          font-weight: normal;
        }

        .smallLink {
            font-size: 10pt;
            margin-left: 5px;
            margin-top: 3px;
            display: inline-block;
        }

        .only-large-screen {
            display: none !important;
        }

        .transcript-color-legend {
            border-style:solid;
            min-width:23px;
            display:inline-block;
            margin-left: 20px;
        }

        @media screen and (min-width: 770px) {
            .only-large-screen {
                display: block !important;
            }
        }

        .results-div {
            width: 100%;
            position: static;
        }

        @media screen and (max-width: 1400px) {
            .results-div {
                width: 85vw;
                position: relative;
            }
        }

    </style>
</head>
<body>
    <div class="ui stackable grid">
      <div class="row">
        <div class="one wide column only-large-screen"></div>
        <div class="ten wide column">
            <div class="ui grid">
                <div class="row">
                    <div class="twelve wide column">
                        <div style="padding:25px 0px">
                            <div>
                                <div style="display:inline-block; min-width: 430px; ">Enter a variant below to see its <a href="https://www.cell.com/cell/pdf/S0092-8674(18)31629-5.pdf" target="_blank">SpliceAI</a> and <a href="https://github.com/tkzeng/Pangolin" target="_blank">Pangolin</a> scores. </div>
                                    <a id="more-details1-button" href="#" onclick="$('#more-details1').show();$('#more-details1-button').hide();">[more details]</a>
                                <div id="more-details1" style="display:none">
                                    <br />
                                    This web interface and the underlying <a href="https://spliceailookup-api.broadinstitute.org/" target="_blank">SpliceAI server</a> are developed and maintained by the <a href="https://the-tgg.org/" target="_blank">TGG</a> at the <a href="https://www.broadinstitute.org/" target="_blank">Broad Institute</a>.<br />
                                    The <a href="https://spliceailookup-api.broadinstitute.org/" target="_blank">SpliceAI server</a> optionally uses pre-computed
                                    scores provided by Illumina for SNVs and small InDels, but runs the <a href="https://github.com/Illumina/SpliceAI" target="_blank">SpliceAI model</a>
                                    directly for scores not available in the pre-computed files - such as for larger InDels, non-default "Max distance" values, etc.<br />
				    <br />
				    NOTE: The server supports not more than several queries per-minute per-user. If you need to batch-process many variants, please use the
				    <a href="https://github.com/broadinstitute/SpliceAI-lookup/blob/master/README.md#local-install" target="_blank">source code</a>
				    to set up your own local instance of the API server or just run the underlying SpliceAI model directly.<br />
                                    <br />
                                    For more information about the SpliceAI model, see <a href="https://doi.org/10.1016/j.cell.2018.12.015" target="_blank">Jaganathan et al, Cell 2019</a> and <a href="https://github.com/Illumina/SpliceAI" target="_blank">github.com/Illumina/SpliceAI</a>,
                                    and this <a href="https://www.youtube.com/watch?v=oJvhj-tYbBI" target="_blank">recorded talk</a> by Kishore Jaganathan.<br />
                                    <br />
				    Since February 2023, <a href="https://github.com/tkzeng/Pangolin" target="_blank">Pangolin scores</a> from [<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02664-4" target="_blank">Zeng & Li 2022</a>] are also shown below the SpliceAI results.
                                </div>
                            </div>
                            <br />
                            <div style="padding-left: 0px; color: gray">
                                <table>
                                    <thead style="text-align: left">
                                        <tr>
                                            <th style="padding-bottom: 20px; min-width:430px">Examples &nbsp; (on hg38): </th>
                                            <th></th>
                                        </tr>
                                    </thead>
                                    <tbody>
                                        <tr>
                                            <td>chr8-140300616-T-G</td>
                                            <td></td>
                                        </tr>
                                        <tr>
                                            <td>chr8 &nbsp; 140300616 &nbsp; T &nbsp; G</td>
                                            <td></td>
                                        </tr>
                                        <tr>
                                            <td style="padding-right: 0px">NM_001089.3(ABCA3):c.875A>T (p.Glu292Val)</td>
                                            <td style="padding-left: 0px">
                                                <a id="more-details2-button" href="#" onclick="$('.more-details2').show();$('#more-details2-button').hide();">[show more examples]</a>
                                            </td>
                                        </tr>
                                        <tr style="display:none;" class="more-details2">
                                            <td>8:140300616 T&gt;G</td>
                                            <td>'chr' <i>prefix is optional</i></td>
                                        </tr>
                                        <tr style="display:none;" class="more-details2">
                                            <td>1:930130:C:G </td>
                                            <td><i>position with overlapping genes</i></td>
                                        </tr>
                                        <tr style="display:none;" class="more-details2">
                                            <td>1-1042601-A-AGAGAG </td>
                                            <td><i>insertion of GAGAG</i></td>
                                        </tr>
                                        <tr style="display:none;" class="more-details2">
                                            <td>1-1042466-GGGC-G</td>
                                            <td><i>deletion of GGC</i></td>
                                        </tr>
                                        <tr style="display:none;" class="more-details2">
                                            <td>NM_000249.4(MLH1):c.116G>A</td>
                                            <td><i>another HGVS exapmle</i></td>
                                        </tr>
                                    </tbody>
                                </table>
                            </div>
                        </div>
<!--
                        <br />
                        <div style="font-size: large; color: darkred">
                          <br />
                          <b>The server is currently down for updates</b><br />
                          <br />
                          Please check back in ~1/2 hour.
                        </div>
                        <br />
                        <br />
-->
                        <div class="ui form" style="width:100%">
                            <div class="ui input" style="padding-bottom:30px; padding-right:8px; width:100%">
                                <input id="search-box" type="text" style="width: 100%; padding: 7px 10px" placeholder="Enter a variant...">
                            </div>
                            <!-- div class="field" style="padding-bottom: 10px;">
                                <textarea id="search-box2" rows="1" style="width: 100%" placeholder="Enter variant..."></textarea>
                            </div -->
                            <div class="ui stackable grid">
                                <div class="row">
                                    <div class="thirteen wide column">
                                        <div class="field" style="padding-right:10px; white-space:nowrap">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Genome version:</b></span>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="hg" value="37" /><label>hg19</label>
                                            </div>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="hg" value="38" checked /><label>hg38</label>
                                            </div>
                                        </div>
                                        <div class="field" style="padding-right:10px; white-space:nowrap">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Score type:</b></span>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="mask" value="1" checked /><label>masked</label>
                                            </div>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="mask" value="0" /><label>raw</label>
                                            </div>
                                            <i style="margin-left:20px" class="question circle outline icon" data-position="right center" data-content="Splicing changes corresponding to strengthening annotated splice sites and weakening unannotated splice sites are typically much less pathogenic than weakening annotated splice sites and strengthening unannotated splice sites. Selecting 'masked' will hide the score for such splicing changes and show 0 instead. Selecting 'raw' will show all scores. SpliceAI developers recommend using 'raw' scores for alternative splicing analysis and 'masked' scores for variant interpretation."></i>
                                        </div>
                                        <div class="field" style="padding-right:10px; white-space:nowrap">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Max distance:</b></span>
                                            <div class="ui input" style="padding-left:15px; padding-bottom:10px; padding-right:8px; width:100px">
                                                <input id="max-distance-input" type="text" value="500" style="padding:7px 10px">
                                            </div>
                                            <i class="question circle outline icon" data-position="right center" data-content="For each variant, SpliceAI looks within a window (+/- 500bp by default) to see how the variant affects the probabilities of different positions being splice acceptors or donors. The distance specified here controls the size of this window. The maximum allowed value is 10,000bp."></i>
                                        </div>
                                        <!--div class="field" style="padding-right:10px; padding-bottom:12px; white-space:nowrap">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Use Illumina's pre-computed scores:</b></span>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="use-precomputed-scores" value="1" /><label>yes</label>
                                            </div>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="use-precomputed-scores" value="0" checked /><label>no</label>
                                            </div>
                                            <i style="margin-left:20px" class="question circle outline icon" data-position="right center" data-content="Illumina provides pre-computed SpliceAI scores for all SNPs and small InDels within Gencode v24 canonical transcripts. If 'yes' is selected here, the server will quickly check these pre-computed files to see if it can avoid the relatively slow step of running the SpliceAI model. Even with 'yes' selected, it will still run the model if it finds that the pre-computed files don't contain the variant. Selecting 'no' will cause the server to always skip the pre-computed files and run the model. The benefit of skipping the pre-computed files is that, although the scores may take longer, they will always be computed for Gencode v38 transcripts and will include ENSG and ENST ids in addition to the gene name."></i>
                                        </div-->
                                    </div>
                                    <div class="three wide column">
                                        <button id="submit-button" class="ui primary button" style="margin-bottom: 15px">Submit</button>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        </div>
        <div class="four wide column">
            <div class="only-large-screen" style="padding:25px 0px">
                <div>
                    <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues" target="_blank">SpliceAI-lookup/issues</a>: issues or feature requests for this website.
                    <a href="https://github.com/illumina/SpliceAI" target="_blank">SpliceAI/issues</a>: issues with the underlying SpliceAI model.<br />
                    <br />
                    <!-- <<div><i>What's New</i></div>
                    <br /> -->
                    <b>Sep 7, 2023</b><br />
                    - added transcript toggling to simplify SpliceAI transcript output<br />
                    - added transcript strand to the results table<br />
                    - added score visualizations for SpliceAI and Pangolin using <a href="https://github.com/bw2/igv.js">custom IGV.js</a> tracks<br />
                    <br />
                    <b>May 22, 2023</b><br />
                    - changed defaults to 'masked' and max distance to 500bp<br />
                    - retired <b>Illumina precomputed scores</b> since they were created using Gencode v24
                    and max distance = 50bp and so can differ from computed scores which are based on the latest Gencode.<br />

		            <br />

                    <a id="more-details3-button" href="#" onclick="$('#more-details3').show();$('#more-details3-button').hide();">[show older updates]</a>
                    <div id="more-details3" style="display:none">
                      <b>May 12, 2023</b><br />
                      - added REF & ALT score columns to show SpliceAI's underlying predictions for each haplotype (using <a href="https://github.com/Illumina/SpliceAI/pull/92">SpliceAI PR by @h-joshi</a>).
                      The &#916;&nbsp; score column is the difference between these two scores.<br />
                      - added support for MNPs (see <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues/1">issue #1</a> & <a href="https://github.com/Illumina/SpliceAI/pull/119">SpliceAI PR by @kdahlo</a>)<br />
                      - updated to <a href="https://www.gencodegenes.org/human/releases.html">Gencode v43</a> (was previously on v42)<br />
                        <b>May 10, 2023</b><br />
                        - fixed bug where "raw" Pangolin scores were shown regardless of whether the user selected "raw" or "masked".<br />
                        <br />

                        <b>Feb 5, 2023</b><br />
                        - show <a href="https://github.com/tkzeng/Pangolin">Pangolin scores</a> in addition to SpliceAI scores.<br />
                        <br />
                        <b>June 2, 2021</b><br />
                        - show RefSeq ids for the subset of Ensembl ENST ids (~30%) that have one or more matching RefSeq NM ids according to Ensembl. See <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues/8">issue #8</a> for details.<br />
                        <br />
                        <b>April 11, 2021</b><br />
                        - show gray background for non-coding transcripts<br />
                        - switch to showing all non-coding transcripts. (Previously, transcripts with Gencode biotypes
                          like <i>lncRNA</i>, <i>processed_pseudogene</i>, <i>processed_transcript</i>, <i>retained_intron</i>, and
                          <i>nonsense_mediated_decay</i> were filtered out if they overlapped <i>protein_coding</i> transcripts).
                          See <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues/6#issuecomment-816942824">issue #6</a> for details.<br />
                    </div>
                </div>
                <br />
                <div>
                    <b>Related web tools:</b><br />
                    <a href="https://liftover.broadinstitute.org">liftover</a>: for variants/positions/intervals (hg19 <=> hg38 <=> T2T)<br />
                    <a href="https://cma-search.broadinstitute.org">CMA search</a>: search OMIM by interval, gene or phenotype<br />
                </div>
            </div>
        </div>
        <div class="one wide column only-large-screen"></div>
      </div>
      <div class="row">
          <div class="one wide column only-large-screen"></div>
          <div class="fourteen wide column">
              <div class="results-div">
                  <div id="error-box" style="color:darkred"></div>
                  <div id="response-box">
                    <table id="scores-box" class="ui stackable table">
                      <thead id="spliceai_header">
                        <tr><th style="background-color: white;" colspan='7'>SpliceAI scores: <a href="https://www.cell.com/cell/pdf/S0092-8674(18)31629-5.pdf" target="_blank"><i class='question circle outline icon' data-position='right center' data-content='SpliceAI scores are computed for each transcript on the server by running the SpliceAI model with the user-selected parameters (ie. max distance).'></i></a></th></tr>
                        <tr>
                          <th>Variant</th>
                          <th>Gene
                              <div style='float:right;'>
                                <span class='transcript-color-legend' style='background-color: var(--CANONICAL_TRANSCRIPT_BACKGROUND_COLOR)!important'>&nbsp;</span> = canonical transcript
                                <span class='transcript-color-legend' style='background-color: var(--NON_CODING_TRANSCRIPT_BACKGROUND_COLOR)!important'>&nbsp;</span> = non-coding transcript
                              </div>
                          </th>
                          <th>&#916;&nbsp; type</th>
                          <th style='white-space: nowrap'>&#916;&nbsp; score<i class='question circle outline icon' data-content='Delta scores range from 0 to 1 and can be interpreted as the probability that the variant affects splicing at any position within a window around it (+/- 500bp by default). In the SpliceAI paper, a detailed characterization is provided for 0.2 (high recall), 0.5 (recommended), and 0.8 (high precision) cutoffs. Consequently, scores of 0.2 or higher appear in green, 0.5 or higher appear in yellow, and 0.8 or higher appear in red.'></i></th>
                          <th style='white-space: nowrap'>pre-mRNA position<i class='question circle outline icon' data-content='For each variant, SpliceAI looks within a window (+/- 500bp by default) to see how the variant affects the probabilities of different positions in the pre-mRNA being splice acceptors or donors. The offsets in this column represent positions with the biggest change in probability within the window. Negative values are to the left of the variant in genomic coords, regardless of transcript strand.'></i></th>
                          <th style='white-space: nowrap'>REF score<i class='question circle outline icon' data-content="SpliceAI's computed probability that the given pre-mRNA position is a splice acceptor or donor based on the reference haplotype sequence. SpliceAI computes delta scores by taking the difference between the REF and ALT score at each position."></i></th>
                          <th style='white-space: nowrap'>ALT score<i class='question circle outline icon' data-content="SpliceAI's computed probability that the given pre-mRNA position is a splice acceptor or donor based on the alternate haplotype sequence. SpliceAI computes delta scores by taking the difference between the REF and ALT score at each position."></i></th>
                        </tr>
                      </thead>
                      <tbody class='transcript_toggle'>
                        <tr><td colspan='7' style='padding: 10px 0px 50px 0px'><div class='ui buttons'><button id='canonicalToggle' class='ui button' onclick='buttonToggle("canonical")'>Canonical</button><div class='or'></div>
                        <span id='diffScoreToggle_span'><button id='diffScoreToggle' class='ui button' onclick='buttonToggle("diffScore")'>Transcripts with Differing Scores</button></span><div class='or'></div>
                        <span id='noncanonicalToggle_span'><button id='noncanonicalToggle' class='ui button' onclick='buttonToggle("noncanonical")'>All Transcripts</button></span></div>
                        </td></tr></tbody>
                      <thead id="pangolin_header">
                        <tr><th style="background-color: white;" colspan='7'>Pangolin scores: <a href='https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02664-4' target='_blank'><i class='question circle outline icon' data-position='right center' data-content='Pangolin scores are computed for each transcript on the server by running the SpliceAI model with the user-selected parameters (ie. max distance).'></i></a></th></tr>
                        <tr>
                          <th>Variant</th>
                          <th>Gene</th>
                          <th>&#916;&nbsp; type</th>
                          <th style='white-space: nowrap'>&#916;&nbsp; score<i class='question circle outline icon' data-content='Delta scores range from 0 to 1 and can be interpreted as the probability that the variant affects splicing at any position within a window around it (+/- 500bp by default). Pangolin does not recommend specific score thresholds; however, here scores of 0.2 or higher appear in green, 0.5 or higher appear in yellow, and 0.8 or higher appear in red.'></i></th>
                          <th style='white-space: nowrap'>pre-mRNA position<i class='question circle outline icon' data-content='For each variant, Pangolin looks within a window (+/- 500bp by default) to see how the variant affects the probabilities of different positions in the pre-mRNA being splice sites. The offsets in this column represent positions with the biggest change in probability within the window. Negative values are upstream (5&#039;) of the variant, and positive are downstream (3&#039;) of the variant.'></i></th>
                          <th style='white-space: nowrap'>REF score<i class='question circle outline icon' data-content="Working on it!"></i></th>
                          <th style='white-space: nowrap'>ALT score<i class='question circle outline icon' data-content="Working on it!"></i></th>
                        </tr>
                      </thead>
                    </table>
                    <div id="igvDiv" style="width:100%; padding-top:10px">
                    </div>
                  </div>
              </div>
          </div>
          <div class="one wide column only-large-screen"></div>
      </div>
    </div>

    <script>
      const VARIANT_RE = new RegExp(
	"^[\\s]*" +
        "(chr)?([0-9XYMTt]{1,2})" +
        "[-\\p{Pd}\\s:]+" +
        "([0-9,]+)" +
        "[-\\p{Pd}\\s:]*" +
        "([ACGT]+)" +
        "[-\\p{Pd}\\s:>]+" +
        "([ACGT]+)",
        'iu'
      )

      const GRCH37_contigs = {
        "1": "NC_000001.10",
        "2": "NC_000002.11",
        "3": "NC_000003.11",
        "4": "NC_000004.11",
        "5": "NC_000005.9",
        "6": "NC_000006.11",
        "7": "NC_000007.13",
        "8": "NC_000008.10",
        "9": "NC_000009.11",
        "10": "NC_000010.10",
        "11": "NC_000011.9",
        "12": "NC_000012.11",
        "13": "NC_000013.10",
        "14": "NC_000014.8",
        "15": "NC_000015.9",
        "16": "NC_000016.9",
        "17": "NC_000017.10",
        "18": "NC_000018.9",
        "19": "NC_000019.9",
        "20": "NC_000020.10",
        "21": "NC_000021.8",
        "22": "NC_000022.10",
        "X": "NC_000023.10",
        "Y": "NC_000024.9",
        "M":  "NC_012920.1",
        "MT": "NC_012920.1",
      }

      const GRCH38_contigs = {
        "1" : "NC_000001.11",
        "2" : "NC_000002.12",
        "3" : "NC_000003.12",
        "4" : "NC_000004.12",
        "5" : "NC_000005.10",
        "6" : "NC_000006.12",
        "7" : "NC_000007.14",
        "8" : "NC_000008.11",
        "9" : "NC_000009.12",
        "10": "NC_000010.11",
        "11": "NC_000011.10",
        "12": "NC_000012.12",
        "13": "NC_000013.11",
        "14": "NC_000014.9",
        "15": "NC_000015.10",
        "16": "NC_000016.10",
        "17": "NC_000017.11",
        "18": "NC_000018.10",
        "19": "NC_000019.10",
        "20": "NC_000020.11",
        "21": "NC_000021.9",
        "22": "NC_000022.11",
        "X" : "NC_000023.11",
        "Y" : "NC_000024.10",
        "M":  "NC_012920.1",
        "MT": "NC_012920.1",
      }

      function parseVariant(variant) {
        const m = variant.match(VARIANT_RE)
        if (m) {
          const chrom = m[2].toUpperCase()
          return {
            chrom: chrom,
            pos: m[3].replace(/,/g, ""),
            ref: m[4],
            alt: m[5],
          }
        }

        return null
      }

      function parseVariantToHGVS(variant, genome_version) {
        const m = variant.match(VARIANT_RE)
        if (m) {
          let chrom = m[2].toUpperCase()
          if (genome_version == "37") {
            chrom = GRCH37_contigs[chrom]
          } else if(genome_version == "38") {
            chrom = GRCH38_contigs[chrom]
          } else {
            throw "Invalid genome_version: " + genome_version
          }

          const pos = parseInt(m[3].replace(/,/g, ""))
          const ref = m[4]
          const alt = m[5]

          if (ref.length > 1 && alt.length > 1) {
            variant = chrom + ":g." + pos + "_" + (pos + ref.length - 1) + "delins" + alt
          } else if (ref.length == alt.length) {
            variant = chrom + ":g." + pos + ref + ">" + alt
          } else if (ref.length > alt.length) {
            //deletion (https://varnomen.hgvs.org/recommendations/DNA/variant/deletion/)
            variant = chrom + ":g." + (pos + 1) + "_" + (pos + ref.length - alt.length) + "del"
          } else if (ref.length < alt.length) {
            //insertion (https://varnomen.hgvs.org/recommendations/DNA/variant/insertion/)
            variant = chrom + ":g." + pos + "_" + (pos + 1) + "ins" + alt.slice(1)
          }

          console.log("Parsed variant", variant)
        }

        return variant
      }

      function variantType(ref, alt) {
        if (ref.length > 1 && alt.length > 1) {
            variant_type = "MNV"
          } else if (ref.length == alt.length) {
            variant_type = "SNV"
          } else if (ref.length > alt.length) {
            variant_type = "DEL"
          } else if (ref.length < alt.length) {
            variant_type = "INS"
          }

          return variant_type
      }

      function formatScore(score) {
          if (score == null || isNaN(score)) {
              return ""
          } else {
              return parseFloat(score).toFixed(2)
          }
      }

      function getScoreStyle(score, isDeltaScore) {
        score = parseFloat(score)
        if (typeof isDeltaScore === 'undefined') {
          return
        }

        const opacity = isDeltaScore ? 1 : 0.3;
        const RED_BACKGROUND = "rgba(252, 207, 184, " + opacity + ")"       // #fccfb8
        const YELLOW_BACKGROUND = "rgba(255, 241, 157, " + opacity + ")" //fff19d
        const GREEN_BACKGROUND = "rgba(205, 255, 215, " + opacity + ")"  //cdffd7

        let result = "style='"
        if(Math.abs(score) >= 0.8) {
            result += "background-color:"+RED_BACKGROUND+";"
        } else if (Math.abs(score) >= 0.5) {
            result += "background-color:"+YELLOW_BACKGROUND+";"
        } else if (Math.abs(score) >= 0.2) {
            result += "background-color:"+GREEN_BACKGROUND+";"
        }
        result += "'"
        return result;
      }

      function getUCSCBrowserUrl(genome_version, chrom, pos) {

        genome_version = genome_version.replace('37', '19')
        chrom = chrom.toUpperCase().replace('CHR', '')
        return "https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg" + genome_version + "&position=chr" + chrom + ":" + pos
      }

      function generateIgvConfig(responseSpliceAI, responsePangolin) {
          //generate IGV config json based on the SpliceAI response

          const chrom = `chr${responseSpliceAI.chrom.replace('chr', '')}`
          const pos = responseSpliceAI.pos
          const ref = responseSpliceAI.ref
          const alt = responseSpliceAI.alt

          let minPos = 1e9
          let maxPos = 0
          for (response of [responseSpliceAI, responsePangolin]) {
              for (x of response.allNonZeroScores) {
                  x["chr"] = chrom
                  x["start"] = x.pos
                  x["end"] = x.pos
                  minPos = Math.min(minPos, x.pos)
                  maxPos = Math.max(maxPos, x.pos)
              }
          }

          // specify IGV tracks and the data to display in them
          const tracks = []

          tracks.push({
              name: "Refseq",
              format: "refgene",
              url: responseSpliceAI.hg == "38" ?
                  "https://hgdownload.soe.ucsc.edu/goldenPath/hg38/database/ncbiRefSeq.txt.gz" :
                  "https://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/ncbiRefSeq.txt.gz",
              indexed: false,
              infoURL: "https://www.ncbi.nlm.nih.gov/gene/?term=$$",
              height: 100,
          })

          tracks.push({
              name: "Variant (?) ",
              type: "vcf",
              description: `This tracks shows the position of the <b>${chrom}:${pos} ${ref}>${alt}</b> variant.`,
              height: 30,
              features: [
                  {
                      chr: chrom,   //see createVCFVariant function in the igv.js repo for the allowed fields
                      pos: pos,
                      start: pos - 1,
                      end: pos,
                      referenceAllele: ref,
                      alternateBases: alt,
                      names: ".", // id in VCF
                      info: {
                          variant: `${chrom}-${pos}-${ref}-${alt}`,
                      }
                  },
              ],
          })

          if (responseSpliceAI) {
              tracks.push({
                  name: `SpliceAI (?)`,
                  description: `<b>Transcript: ${responseSpliceAI.allNonZeroScoresTranscriptId}</b> (${responseSpliceAI.allNonZeroScoresStrand == '-' ? 'minus' : 'plus'} strand)<br />
                                SpliceAI predictions of splice acceptor & donor sites within the +/- ${responseSpliceAI.distance}bp window<br />
                                around <b>${chrom}:${pos} ${ref}>${alt}</b>. This track shows markers that represent the scores for the<br/>
                                <span style="color:#0000B4"><b>reference sequence</b></span> (without the variant) and the<br />
                                <span style="color:#05d0d2"><b>alternate sequence</b></span> (with the variant). <br />
                                A marker is shown at all positions where a score is ≥ 0.01. <br/>
                                SpliceAI outputs separate predictions for whether a given position is a splice acceptor and/or a splice donor, so<br/>
                                <b>A</b>'s represent acceptor probability scores, and <br/>
                                <b>D</b>'s represent donor probability scores.<br/>
                                The letters are shown up-side-down when the alternate sequence score is lower than the reference sequence score.<br/>
                                Numerical labels show the scores predicted from the reference sequence.
                                `,
                  height: 100,
                  rawOrDelta: "raw",
                  tool: "SpliceAI",
                  type: "spliceprediction",
                  features: responseSpliceAI.allNonZeroScores,
                  strand: responseSpliceAI.allNonZeroScoresStrand,
              })
              tracks.push({
                  name: `SpliceAI (?)`,
                  description: `<b>Transcript: ${responseSpliceAI.allNonZeroScoresTranscriptId}</b> (${responseSpliceAI.allNonZeroScoresStrand == '-' ? 'minus' : 'plus'} strand)<br />
                                SpliceAI delta score predictions for gain and loss of splice acceptors and donors within a <br />
                                +/- ${responseSpliceAI.distance}bp window around <b>${chrom}:${pos} ${ref}>${alt}</b>.<br />
                                This track shows markers at all positions where the |delta score| is ≥ 0.01. <br/>
                                <b>Bar colors:</b> <b style="color:#FF0000">red</b> is for delta scores ≥ 0.8,
                                <b style="color:#FFCB1F">yellow</b> is for delta scores ≥ 0.5,
                                <b style="color:#1fb839">green</b> is for delta scores ≥ 0.2<br />
                                <b>A</b>'s represent acceptor probability scores<br/>
                                <b>D</b>'s represent donor probability scores<br/>
                                `,
                  height: 200,
                  rawOrDelta: "delta",
                  tool: "SpliceAI",
                  type: "spliceprediction",
                  features: responseSpliceAI.allNonZeroScores,
                  strand: responseSpliceAI.allNonZeroScoresStrand,
              })
          }

          if (responsePangolin) {
              tracks.push({
                  name: `Pangolin (?)`,
                  description: `<b>Gene: ${responsePangolin.allNonZeroScoresTranscriptId}</b> (${responsePangolin.allNonZeroScoresStrand == '-' ? 'minus' : 'plus'} strand)<br />
                                Pangolin delta score predictions for gain and loss of splice sites at positions<br />
                                within a +/- ${responsePangolin.distance}bp window around <b>${chrom}:${pos} ${ref}>${alt}</b>.<br />
                                This track shows markers at all positions where the delta score is ≥ 0.1. <br/>
                                <b>Bar colors:</b> <b style="color:#FF0000">red</b> is for delta scores ≥ 0.8,
                                <b style="color:#FFCB1F">yellow</b> is for delta scores ≥ 0.5,
                                <b style="color:#1fb839">green</b> is for delta scores ≥ 0.2<br />
                                `,
                  height: 200,
                  rawOrDelta: "delta",
                  tool: "Pangolin",
                  type: "spliceprediction",
                  features: responsePangolin.allNonZeroScores,
                  strand: responsePangolin.allNonZeroScoresStrand,
              })
          }

          const gencodeTrackPath = responseSpliceAI.hg == "37" ?
              "gs://tgg-viewer/ref/GRCh37/gencode_v44/knownGene_v44.hg19.sorted.txt.gz" :
              "gs://tgg-viewer/ref/GRCh38/gencode_v44/knownGene_v44.hg38.sorted.txt.gz"

          tracks.push({
                  name: "Gencode v44",
                  format: 'refgene',
                  url: gencodeTrackPath,
                  indexURL: `${gencodeTrackPath}.tbi`,
                  indexed: true,
                  searchable: true,
                  height: 350,
                  visibilityWindow: -1,
                  order: 1000001,
                  displayMode: 'EXPANDED',
                  color: 'rgb(76,171,225)',
          })


          for (const minScore of [0.2, 0.5]) {
              for (const splicePredictionType of ["loss", "gain"]) {
                      tracks.push(
                      {
                          name: `SpliceAI: A or D ${splicePredictionType} ≥ ${minScore} (?)`,
                          description: `This track visualizes Illumina's precomputed SpliceAI score tables for SNVs and small INDELs.<br/>
                             Each genomic location where an SNV or INDEL variant has a SpliceAI score ≥ ${minScore} is shown as the origin of an arrow.<br />
                             The arrow points to the location where that variant most likely causes acceptor or donor ${splicePredictionType}.
                             Clicking on the arrow will display additional information.`,
                          type: "spliceJunctions",
                          height: 100,
                          url: `gs://tgg-viewer/ref/GRCh38/spliceai/spliceai_scores.raw.snps_and_indels.hg38.filtered.sorted.score_${minScore}.splice_${splicePredictionType}.bed.gz`,
                          indexURL: `gs://tgg-viewer/ref/GRCh38/spliceai/spliceai_scores.raw.snps_and_indels.hg38.filtered.sorted.score_${minScore}.splice_${splicePredictionType}.bed.gz.tbi`,
                      }
                  )
              }
          }


          for (const [filenamePrefix, tissue, sampleCount] of [
              ["GTEX_muscle.803_samples", "muscle", 803],
              ["GTEX_blood.755_samples", "blood", 755],
              ["GTEX_fibs.504_samples", "fibroblasts", 504],
              ["GTEX_lymphocytes.174_samples", "lymphocytes", 174],
              ["GTEX_brain_cortex.255_samples", "brain cortex", 255],
              //["GTEX_frontal_cortex.209_samples", "frontal cortex", 803],
          ]) {
              tracks.push({
                  name: `GTEx ${tissue}: ${sampleCount} samples (?)`,
                  description: `This track shows the combined splice junctions from all ${sampleCount} ${tissue} samples available in GTEx v8.
                                Splice junctions are labeled with the total number of RNA-seq reads that supported the junction, summed across the ${sampleCount} samples.
                                The coverage track shows the total number of RNA-seq reads that overlap each position, summed across all samples and divided by ${sampleCount}.`,
                  type: "merged",
                  height: 100,
                  tracks: [{
                      type: "wig",
                      format: "bigwig",
                      url: `gs://tgg-viewer/ref/GRCh38/gtex_v8/${filenamePrefix}.normalized.bigWig`,
                  }, {
                      type: 'spliceJunctions',
                      format: 'bed',
                      minJunctionEndsVisible: 1,
                      colorBy: 'strand',
                      minTotalReads: 3,
                      url: `gs://tgg-viewer/ref/GRCh38/gtex_v8/${filenamePrefix}.junctions.bed.gz`,
                      indexURL: `gs://tgg-viewer/ref/GRCh38/gtex_v8/${filenamePrefix}.junctions.bed.gz.tbi`,
                  }],
              })
          }

          let locusMargin = maxPos - minPos < 200 ? 15 : 50

          // igv.js reference genome specs are copied from https://igv.org/genomes/genomes.json
          // These must be specified explicitly rather than just setting the reference to "hg19" or "hg38" so that
          // the RefSeq gene track can be the first track (as described in https://github.com/igvteam/igv.js/issues/1518)
          let reference = responseSpliceAI.hg == "37" ? {
              "id": "hg19",
              "name": "Human (GRCh37/hg19)",
              "fastaURL": "https://igv-genepattern-org.s3.amazonaws.com/genomes/seq/hg19/hg19.fasta",
              "indexURL": "https://igv-genepattern-org.s3.amazonaws.com/genomes/seq/hg19/hg19.fasta.fai",
              "cytobandURL": "https://igv-genepattern-org.s3.amazonaws.com/genomes/seq/hg19/cytoBand.txt",
              "aliasURL": "https://s3.amazonaws.com/igv.org.genomes/hg19/hg19_alias.tab",
              "chromosomeOrder": "chr1, chr2, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chr10, chr11, chr12, chr13, chr14, chr15, chr16, chr17, chr18, chr19, chr20, chr21, chr22, chrX, chrY",
              "tracks": tracks,
          } : {
              "id": "hg38",
              "name": "Human (GRCh38/hg38)",
              "fastaURL": "https://igv-genepattern-org.s3.amazonaws.com/genomes/seq/hg38/hg38.fa",
              "indexURL": "https://igv-genepattern-org.s3.amazonaws.com/genomes/seq/hg38/hg38.fa.fai",
              "cytobandURL": "https://s3.amazonaws.com/igv.org.genomes/hg38/annotations/cytoBandIdeo.txt.gz",
              "aliasURL": "https://s3.amazonaws.com/igv.org.genomes/hg38/hg38_alias.tab",
              "chromosomeOrder": "chr1, chr2, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chr10, chr11, chr12, chr13, chr14, chr15, chr16, chr17, chr18, chr19, chr20, chr21, chr22, chrX, chrY",
              "tracks": tracks,
          }

          return {
              reference: reference,
              locus: `${chrom}:${minPos - locusMargin}-${maxPos + locusMargin}`,
          }
      }

      function generateResultsTable(response) {
        let parsed_scores_or_errors = []
        // responses = responses.filter(r => r) // what does this filter do?
        const is_pangolin = response.source && response.source.startsWith("pangolin")
        const is_spliceai = response.source && response.source.startsWith("spliceai")
        if (response.error) {
            parsed_scores_or_errors.push({
                'error': response.error,
                'source': response.source,
                'hg': response.hg,
                'variant': response.variant,
                'sort_key': 0,
            })
            return
        }

        response.scores.forEach(function(scores) {
            try {
                scores["url"] = getUCSCBrowserUrl(response.hg, response.chrom, response.pos)
            } catch(e) {
                scores["url"] = "#"
            }

            scores['source'] = response.source
            scores['is_pangolin'] = is_pangolin
            scores['hg'] = response.hg
            scores['variant'] = response.variant

            // compute sort key
            const splice_prediction_tool = is_pangolin ? 1 : 0;

            const gene_name_fields = (scores["NAME"] || "").split("---")
            const gene_name = gene_name_fields[0]
            const transcript_id = gene_name_fields.length >= 3 && gene_name_fields[2]
            const is_canonical_transcript = (gene_name_fields.length >= 4 && gene_name_fields[3] === "yes") ? 0 : 1
            const transcript_type = gene_name_fields.length >= 5 && gene_name_fields[4]
            const is_protein_coding = !transcript_type || transcript_type == "protein_coding" ? 0 : 1
            const refseq_transcript_id_is_in_query = gene_name_fields.length >= 6 && gene_name_fields[5] && gene_name_fields[5].split(",").some(refseq_transcript_id => response.raw.includes(refseq_transcript_id))
            const contains_hgvs_transcript_id = response.raw.includes(transcript_id) || refseq_transcript_id_is_in_query ? 0 : 1;

            scores['sort_key'] = `${splice_prediction_tool}${contains_hgvs_transcript_id}`
            scores['sort_key'] += `${is_protein_coding}${is_canonical_transcript}`
            scores['sort_key'] += `---${gene_name}`

            parsed_scores_or_errors.push(scores)
        })

        //sort by (gene name, is canonical transcript, is non-coding transcriipt)
        parsed_scores_or_errors = _.chain(parsed_scores_or_errors).sortBy('sort_key').value()

        return parsed_scores_or_errors.map(function(scores_or_error, index) {

                let scores = scores_or_error
                const variant_ucsc_url = scores.url
                const gnomAD_version = scores.hg === "38"? "gnomad_r3" : "gnomad_r2_1"
                const variant_gnomAD_url = "https://gnomad.broadinstitute.org/variant/" + scores.variant + "?dataset=" + gnomAD_version
                const gene_name_fields = (scores.NAME || "").split("---")
                const gene_name = gene_name_fields[0].split(".")[0]  // remove gene version number  TODO remove this after switching Pangolin to use regular gene names
                const strand = scores.STRAND == "-" ? "minus" : "plus"
                const is_canonical_transcript = gene_name_fields.length >= 4 && gene_name_fields[3] === "yes"
                const transcript_type = gene_name_fields.length >= 5 && gene_name_fields[4]
                const is_protein_coding = !transcript_type || transcript_type == "protein_coding"
                const refseq_transcript_ids = gene_name_fields.length >= 6 && gene_name_fields[5] && gene_name_fields[5].split(",")

                let gene_and_transcript_id = ""
                if (gene_name_fields.length >= 3 ) {
                    const gene_id_link = `<a href="https://useast.ensembl.org/Homo_sapiens/Gene/Summary?g=${gene_name_fields[1].split(".")[0]}" target="_blank">${gene_name_fields[1]}</a>`
                    const transcript_id_link = ` / <a href="https://useast.ensembl.org/Homo_sapiens/Transcript/Summary?t=${gene_name_fields[2].split(".")[0]}" target="_blank">${gene_name_fields[2]}</a>`
                    const canonical_transcript_text = is_canonical_transcript ? '<div class="smallLink">canonical transcript</div>' : ''
                    const biotype_link = `<div class="smallLink"><a href="https://www.gencodegenes.org/pages/biotypes.html" target="_blank">${transcript_type.replace(/_/g, " ")}</a></div>`
                    const refseq_transcript_id_link = refseq_transcript_ids ? (" /" + ` <a href="https://www.ncbi.nlm.nih.gov/search/all/?term=${refseq_transcript_ids[0]}" target="_blank">${refseq_transcript_ids[0]}</a>`) : ''
                    gene_and_transcript_id = ` <div class="smallLink">(${gene_id_link}${transcript_id_link}${refseq_transcript_id_link})</div><br /><br class='only-large-screen' />${biotype_link}${canonical_transcript_text} <div class="smallLink"> (${strand} strand)</div>`
                } else {
                    gene_and_transcript_id = ` <div class="smallLink"> (${strand} strand)</div>`
                }
                //round to 2 decimal places
                if (scores.is_pangolin) {
                    scores.DS_SL = Math.round(scores.DS_SL * 100) / 100
                    scores.DS_SG = Math.round(scores.DS_SG * 100) / 100
                    scores.SL_REF = Math.round(scores.SL_REF * 100) / 100
                    scores.SL_ALT = Math.round(scores.SL_ALT * 100) / 100
                    scores.SG_REF = Math.round(scores.SG_REF * 100) / 100
                    scores.SG_ALT = Math.round(scores.SG_ALT * 100) / 100
                } else {
                    scores.DS_AG = Math.round(scores.DS_AG * 100) / 100
                    scores.DS_AL = Math.round(scores.DS_AL * 100) / 100
                    scores.DS_DL = Math.round(scores.DS_DL * 100) / 100
                    scores.DS_DG = Math.round(scores.DS_DG * 100) / 100
                    scores.DS_AL_REF = Math.round(scores.DS_AL_REF * 100) / 100
                    scores.DS_AL_ALT = Math.round(scores.DS_AL_ALT * 100) / 100
                    scores.DS_DL_REF = Math.round(scores.DS_DL_REF * 100) / 100
                    scores.DS_DL_ALT = Math.round(scores.DS_DL_ALT * 100) / 100
                    scores.DS_AG_REF = Math.round(scores.DS_AG_REF * 100) / 100
                    scores.DS_AG_ALT = Math.round(scores.DS_AG_ALT * 100) / 100
                    scores.DS_DG_REF = Math.round(scores.DS_DG_REF * 100) / 100
                    scores.DS_DG_ALT = Math.round(scores.DS_DG_ALT * 100) / 100
                }

                result_classes = (scores.is_pangolin ? "pangolin" : (is_canonical_transcript ? "spliceai canonical" : (is_protein_coding ?  "spliceai noncanonical coding": "spliceai noncanonical noncoding")))

                let result= "<tbody id='transcript_" + index + "' class='" + result_classes + "'><tr class='upperBorder'>" +
                  "<td rowspan='" + (scores.is_pangolin ? 2: 4) + "' style='vertical-align: top'>" +
                  "<div style='margin-right:10px; display: inline-block'>" + scores.variant + "</div><br class='only-large-screen'/><br class='only-large-screen'/>" +
                    "<a href='"+variant_ucsc_url+"' class='smallLink' target='_blank'>UCSC</a>, " +
                    "<a href='"+variant_gnomAD_url+"' class='smallLink' target='_blank'>gnomAD</a>" +
                  "</td>"+
                  "<td rowspan='" + (scores.is_pangolin ? 2: 4) + "' style='vertical-align: top'>" +
                    "<div style='margin-right:10px; display: inline-block'>"+gene_name+(gene_and_transcript_id || '') +"</div><br class='only-large-screen'/><br class='only-large-screen'/>" +
                    "<a href='https://www.omim.org/search?search="+gene_name+"' class='smallLink' target='_blank'>OMIM</a>, " +
                    "<a href='https://gtexportal.org/home/gene/"+gene_name+"' class='smallLink' target='_blank'>GTEx</a>, " +
                    "<a href='https://gnomad.broadinstitute.org/gene/"+gene_name+"?dataset="+gnomAD_version+"' class='smallLink' target='_blank'>gnomAD</a>, " +
                    "<a href='https://search.clinicalgenome.org/kb/genes?page=1&size=25&search="+gene_name+"' class='smallLink' target='_blank'>ClinGen</a>, <br class='only-large-screen'/>" +
                    "<a href='https://useast.ensembl.org/Homo_sapiens/Gene/Summary?g="+gene_name+"' class='smallLink' target='_blank'>Ensembl</a>, " +
                    "<a href='https://decipher.sanger.ac.uk/gene/"+gene_name+"' class='smallLink' target='_blank'>Decipher</a>, " +
                    "<a href='https://www.genecards.org/cgi-bin/carddisp.pl?gene="+gene_name+"' class='smallLink' target='_blank'>GeneCards</a>" +
                  "</td>" +
                    (scores.is_pangolin ?
                            "<td>Splice Loss</td><td "+getScoreStyle(scores.DS_SL, true)+">" + formatScore(scores.DS_SL) +"</td><td>" + (scores.DS_SL == 0 ? "" : scores.DP_SL+" bp") + "</td><td "+getScoreStyle(scores.SL_REF)+">" + formatScore(scores.SL_REF) + "</td><td "+getScoreStyle(scores.SL_ALT)+">" + formatScore(scores.SL_ALT) + "</td>"+
                            "</tr><tr>"+
                            "<td>Splice Gain</td><td "+getScoreStyle(scores.DS_SG, true)+">" + formatScore(scores.DS_SG) +"</td><td>" + (scores.DS_SG == 0 ? "" : scores.DP_SG+" bp") + "</td><td "+getScoreStyle(scores.SG_REF)+">" + formatScore(scores.SG_REF) + "</td><td "+getScoreStyle(scores.SG_ALT)+">" + formatScore(scores.SG_ALT) + "</td>"
                        :
                            "<td>Acceptor Loss</td><td "+getScoreStyle(scores.DS_AL, true)+">" + formatScore(scores.DS_AL) +"</td><td>" + ((scores.DS_AL == 0 && scores.DS_AL_REF == 0 && scores.DS_AL_ALT == 0)? "" : scores.DP_AL+" bp") + "</td><td "+getScoreStyle(scores.DS_AL_REF)+">" + formatScore(scores.DS_AL_REF) + "</td><td "+getScoreStyle(scores.DS_AL_ALT)+">" + formatScore(scores.DS_AL_ALT) + "</td>"+
                            "</tr><tr>"+
                            "<td>Donor Loss</td><td "+getScoreStyle(scores.DS_DL, true)+">" + formatScore(scores.DS_DL) +"</td><td>" + ((scores.DS_DL == 0 && scores.DS_DL_REF == 0 && scores.DS_DL_ALT == 0)? "" : scores.DP_DL+" bp") + "</td><td "+getScoreStyle(scores.DS_DL_REF)+">" + formatScore(scores.DS_DL_REF) + "</td><td "+getScoreStyle(scores.DS_DL_ALT)+">" + formatScore(scores.DS_DL_ALT) + "</td>"+
                            "</tr><tr>"+
                            "<td>Acceptor Gain</td><td "+getScoreStyle(scores.DS_AG, true)+">" + formatScore(scores.DS_AG) +"</td><td>" + ((scores.DS_AG == 0 && scores.DS_AG_REF == 0 && scores.DS_AG_ALT == 0)? "" : scores.DP_AG+" bp") + "</td><td "+getScoreStyle(scores.DS_AG_REF)+">" + formatScore(scores.DS_AG_REF) + "</td><td "+getScoreStyle(scores.DS_AG_ALT)+">" + formatScore(scores.DS_AG_ALT) + "</td>"+
                            "</tr><tr>"+
                            "<td>Donor Gain</td><td "+getScoreStyle(scores.DS_DG, true)+">" + formatScore(scores.DS_DG) +"</td><td>" + ((scores.DS_DG == 0 && scores.DS_DG_REF == 0 && scores.DS_DG_ALT == 0)? "" : scores.DP_DG+" bp") + "</td><td "+getScoreStyle(scores.DS_DG_REF)+">" + formatScore(scores.DS_DG_REF) + "</td><td "+getScoreStyle(scores.DS_DG_ALT)+">" + formatScore(scores.DS_DG_ALT) + "</td>"
                    ) +
                  "</tr></tbody>"

                return result
              }).join("") +
          "</table>"
      }

      function buttonToggle(category) {
        $("#canonicalToggle").removeClass("primary")
        $("#diffScoreToggle").removeClass("primary")
        $("#noncanonicalToggle").removeClass("primary")

        if (category == "canonical") {
          $(".spliceai.noncanonical").hide()
          $("#canonicalToggle").addClass("primary")
        } else if (category == "diffScore") {
          $(".spliceai.noncanonical").hide()
          $(".spliceai.diffScore").show()
          $("#diffScoreToggle").addClass("primary")
        } else if (category == "noncanonical") {
          $(".spliceai.noncanonical").show()
          $("#noncanonicalToggle").addClass("primary")
        }
      }

      $.urlParam = function (name) {
        const results = new RegExp('[\?&#]?' + name + '=([^&#]*)').exec(window.location.hash);
        return (results !== null) ? decodeURIComponent(results[1]) || 0 : false;
      }

      function getSpliceAI(row, request) {
        scores = []
        positions = []
        Array.from(row.children).forEach(function(item, index) {
          if (item.classList.contains("upperBorder")) {
            scores[index] = item.children[3].innerText
            positions[index] = item.children[4].innerText
          } else {
            scores[index] = item.children[1].innerText
            positions[index] = item.children[2].innerText
          }})
        if (request == "scores") {
          return scores
        } else if (request == "positions") {
          return positions
        } else {
          return [scores, positions]
        }
      }

      function toggleSetUp(canonical) {
        $(".noncanonical").hide()
        $("#canonicalToggle").addClass("primary")
        $("#noncanonicalToggle").removeClass("primary")
        $("#diffScoreToggle").removeClass("primary")
        $("#noncanonicalToggle").removeClass("disabled")
        $("#diffScoreToggle").removeClass("disabled")
        $("#noncanonicalToggle_span").removeClass()
        $("#diffScoreToggle_span").removeClass()

        $("#noncanonicalToggle_span").popup('destroy');
        $('#diffScoreToggle_span').popup('destroy');

        if (document.getElementsByClassName("noncanonical").length == 0) {
          $("#noncanonicalToggle").addClass("disabled")
          $("#noncanonicalToggle_span").addClass("singleTranscript");
          $("#diffScoreToggle").addClass("disabled");
          $("#diffScoreToggle_span").addClass("singleTranscript");

          $(".singleTranscript")
            .popup({
              content : 'Only one transcript reported',
              position: 'top center'
            })
        } else {
          Array.from(document.getElementsByClassName("noncanonical")).forEach(function(item) {
            Array.from(item.children).forEach(function(entry, index) {
              if(!item.classList.contains("diffScore")) {
                if (entry.classList.contains("upperBorder")) {
                  if (entry.children[3].innerText != canonical[index]) {
                    $("#" + item.id).addClass("diffScore")
                  }
                } else if (entry.children[1].innerText != canonical[index]) {
                  $("#" + item.id).addClass("diffScore")
                }}
              })
          });

          if (document.getElementsByClassName("diffScore").length == 0) {
            $("#diffScoreToggle").addClass("disabled");
            $("#diffScoreToggle_span")

            $('#diffScoreToggle_span')
              .popup({
                content : 'All transcripts have the same scores',
                position: 'top center'
              })
          }
        }
      }

      $(document).ready(function() {
        $('.ui.radio.checkbox').checkbox()
        $('.question').popup()
        $("#search-box").focus()
        $("#response-box").hide()

        $("#search-box,#max-distance-input").keydown(function (event) {
          if ((event.keyCode || event.which) == 13) {
            $("#submit-button").click()
          }
        });

        $("#submit-button").click(function() {
          const variant = $("#search-box").val().trim()
          if (!variant) {
            return
          }

          const genome_version = $("input[name='hg']:checked").val().trim()
          const mask = $("input[name='mask']:checked").val().trim()
          const max_distance = $("#max-distance-input").val().trim()
          //const use_precomputed_scores = $("input[name='use-precomputed-scores']:checked").val().trim()
          const variant_hgvs = parseVariantToHGVS(variant, genome_version)

          $("#error-box").hide()
          $("#response-box").hide()
          $(".spliceai").remove()
          $(".pangolin").remove()

          $("#submit-button").addClass(["loading", "disabled"])

          let ensembl_api_url_prefix = ""
          if (genome_version == "37") {
            ensembl_api_url_prefix = "grch37."
          }

          $.getJSON(
            "https://" + ensembl_api_url_prefix + "rest.ensembl.org/vep/human/hgvs/" + variant_hgvs + "?content-type=application/json&vcf_string=1"
          ).catch(function(error) {
            let errorText = error.responseText
            if (errorText) {
                try {
                    errorText = JSON.parse(errorText).error
                } catch (e) {
                    console.warn(e)
                }
            }
            if (errorText) {
                const refAlleleError = errorText.match(new RegExp("[(]([ACGTRYSWKMBDHVN]+)[)] does not match reference allele given by HGVS notation"))
                if (refAlleleError) {
                    errorText = variant + " has unexpected reference allele. The hg" + genome_version + " reference allele should be: " + refAlleleError[1]
                }

                errorText = "Ensembl API error: " + errorText
                console.warn(errorText)
            }

            throw { responseText: errorText }
          }).then(function(response) {
            console.log("Ensembl API response", response)
            if (!response || !response[0]) {
              throw { responseText: "Unable to parse Ensembl API repsonse: " + response }
            }

            const vcf_string = (response[0].vcf_string || "---").split("-")
            const chrom = vcf_string[0]
            const pos = vcf_string[1]
            const ref = vcf_string[2]
            const alt = vcf_string[3]
            const variant_type = variantType(ref, alt)
            const most_severe_consequence = (response[0].most_severe_consequence || "").replace(/_/g, " ")

            console.log("chrom:", chrom, "pos:", pos, "ref:", ref, "alt:", alt)
            console.log("Most severe consequence:", most_severe_consequence)

            // get SpliceAI and Pangolin scores from API endpoints in parallel
            //const base_api_url = location.hostname == "" ? "http://localhost:8080" : "https://spliceailookup-api.broadinstitute.org"
            const getJSON_promises = ["spliceai", "pangolin"].map((endpoint) => {
                const promise = $.getJSON(`${base_api_url}/${endpoint}/`,
                    {
                        hg: genome_version,
                        distance: max_distance,
                        mask: mask,
                        //precomputed: use_precomputed_scores,
                        variant: chrom + '-' + pos + '-' + ref + '-' + alt,
                        raw: variant,
                    }
                )
                return promise
            })

            return Promise.all(getJSON_promises)

          }).catch(function(error) {
              console.log("Error getting scores from SpliceAI or Pangolin APIs:", error)
              throw { responseText: error.errorText }

          }).then(function(response) {
              console.log("spliceai API response was:", response[0])
              console.log("pangolin API response was:", response[1])

              $("#error-box").hide()
              $("#response-box").show()
              if (response[0]) {  // handle timeouts or other errors with the SpliceAI api call
                  $("#spliceai_header").after(generateResultsTable(response[0]))

                  const canonicalTranscriptRow = document.getElementsByClassName("canonical")[0]
                  if (canonicalTranscriptRow) {
                      const canonical = getSpliceAI(canonicalTranscriptRow)
                      toggleSetUp(canonical[0])
                  }

                  const igvConfig = generateIgvConfig(response[0], response[1])
                  if (!window.igvBrowserCreated) {
                      // create igv and load tracks after 1st variant scores received.
                      window.igvBrowserCreated = true
                      igv.createBrowser(document.getElementById("igvDiv"), igvConfig).then((browser) => {
                          window.igvBrowser = browser
                      })
                  } else if(window.igvBrowser) {
                      // page already initialized, and the user submitted another variant. Update IGV tracks.
                      console.log("Updating igv config to", igvConfig)
                      window.igvBrowser.loadSessionObject(igvConfig)
                  }
              }

              if (response[1]) {  // handle timeouts or other errors with the Pangolin api call
                  $("#pangolin_header").after(generateResultsTable(response[1]))
              }

              // set url
              window.location.hash = "#" + $.param({
                  variant: variant,
                  hg: genome_version,
                  distance: max_distance,
                  mask: mask,
              })

          }).catch(function(error) {

            $("#response-box").hide()
            let errorText = error.responseText
            try {
              errorText = JSON.parse(errorText).error
            } catch(e) {
            }

            if (!errorText) {
                errorText = "ERROR: Unable to reach the SpliceAI-lookup API server. Either the API server is down or your internet isn't working."
            }
            console.warn(errorText)
            $("#error-box").html(errorText)
            $("#error-box").show()
          }).always(function() {
            $("#submit-button").removeClass(["loading", "disabled"]);
            $('.question').popup()
          })
        })

        const hgFromUrl = $.urlParam('hg')
        if (hgFromUrl) {
          $("input[name='hg'][value='"+hgFromUrl+"']").prop("checked", true)
        }

        const maxDistanceFromUrl = $.urlParam('distance')
        if (maxDistanceFromUrl) {
            $("#max-distance-input").val(maxDistanceFromUrl)
        }

        const maskFromUrl = $.urlParam('mask')
        if (maskFromUrl) {
          $("input[name='mask'][value='"+maskFromUrl+"']").prop("checked", true)
        }

        /*
        const precomputedFromUrl = $.urlParam('precomputed')
        if (precomputedFromUrl) {
          $("input[name='use-precomputed-scores'][value='"+precomputedFromUrl+"']").prop("checked", true)
        }
        */

        const variantFromUrl = $.urlParam('variant')
        if (variantFromUrl) {
          $("#search-box").val(variantFromUrl)
          $("#submit-button").click()
        }
      })

    </script>
</body>
</html>
