<html>
<head>
    <script>
        const isBetaSite = !location.origin.includes("spliceailookup.broadinstitute.org")

        let baseApiUrl
        if (location.origin.includes("broadinstitute.")) {
            baseApiUrl = "https://spliceailookup-api.broadinstitute.org"
        } else if (location.origin && location.origin.startsWith("http")) {
            baseApiUrl = location.origin
        } else {
            baseApiUrl = "http://localhost:8080"
        }

        console.log(isBetaSite ? "BETA site" : "Production site.", "Using SpliceAI-lookup API url:", baseApiUrl)
    </script>
    <meta content="width=device-width, initial-scale=1" charset="utf-8" name="viewport"/>
    <title>SpliceAI Lookup</title>
    <link rel="shortcut icon" href="/icon.png">

    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.css"/>
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.js"></script>
    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/underscore@1.12.0/underscore-min.js"></script>

    <script type="text/javascript" src="https://storage.googleapis.com/tgg-viewer/igv.js-bw2/dist/igv.min.js"></script>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-172722632-1"></script>
    <script>
        window.dataLayer = window.dataLayer || [];
        function gtag(){dataLayer.push(arguments);}
        gtag('js', new Date());
        gtag('config', 'UA-172722632-1');
    </script>
    <style>
        :root {
            --MAIN_TRANSCRIPT_BACKGROUND_COLOR: #eaf3ff;
            --CODING_TRANSCRIPT_BACKGROUND_COLOR: white;
            --NON_CODING_TRANSCRIPT_BACKGROUND_COLOR: #e5e5e5;
        }

        .coding-transcript {
            background-color: var(--CODING_TRANSCRIPT_BACKGROUND_COLOR);
        }

        .non-coding-transcript {
            background-color: var(--NON_CODING_TRANSCRIPT_BACKGROUND_COLOR);
        }

        .main-transcript {
            background-color: var(--MAIN_TRANSCRIPT_BACKGROUND_COLOR);
        }

        body {
            overflow-x: scroll;
        }

        body, .ui.form, label {
            font-size: 11pt !important;
        }

        td {
            padding: 5px 15px;
        }

        #pangolin-header th {
            border-top: 1px solid rgba(34, 36, 38, .15);
        }

        table {
            font-size: 11pt !important;
            border-collapse: collapse;
            border-left: 0px !important;
            border-right: 0px !important;
            width: 100%;
        }

        th div {
            font-weight: normal;
        }

        .small-link {
            font-size: 10pt;
            margin-left: 5px;
            margin-top: 3px;
            display: inline-block;
        }

        .only-large-screen {
            display: none !important;
        }

        .transcript-color-legend {
            border-style: solid;
            min-width: 23px;
            display: inline-block;
            margin-left: 20px;
        }

        @media screen and (min-width: 770px) {
            .only-large-screen {
                display: block !important;
            }
        }

        .results-div {
            width: 100%;
            position: static;
        }

        @media screen and (max-width: 1400px) {
            .results-div {
                width: 85vw;
                position: relative;
            }
        }

    </style>
</head>
<body>
<div class="ui stackable grid">
    <div class="row">
        <div class="one wide column only-large-screen"></div>
        <div class="ten wide column">
            <div class="ui grid">
                <div class="row">
                    <div class="twelve wide column">
                        <div style="padding:25px 0px">
                            <div>
                                <div style="display:inline-block; min-width: 700px; ">Enter a variant below to see its
                                    <a href="https://www.cell.com/cell/pdf/S0092-8674(18)31629-5.pdf" target="_blank">SpliceAI</a>
                                    and <a href="https://github.com/tkzeng/Pangolin" target="_blank">Pangolin</a> scores <span id="other-version-of-this-site-link"></span>
                                </div>
                                <a id="more-details1-button" href="#" onclick="$('#more-details1').show();$('#more-details1-button').hide();">[more details]</a>
                                <div id="more-details1" style="display:none">
                                    <br/>
                                    This web interface and the underlying <a href="https://spliceailookup-api.broadinstitute.org/" target="_blank">SpliceAI
                                    server</a> are developed and maintained by the <a href="https://the-tgg.org/" target="_blank">TGG</a> at the <a
                                        href="https://www.broadinstitute.org/" target="_blank">Broad Institute</a>.<br/>
                                    The <a href="https://spliceailookup-api.broadinstitute.org/" target="_blank">SpliceAI
                                    server</a> optionally uses pre-computed scores provided by Illumina for SNVs and small InDels, but runs the <a
                                        href="https://github.com/Illumina/SpliceAI" target="_blank">SpliceAI model</a>
                                    directly for scores not available in the pre-computed files - such as for larger
                                    InDels, non-default "Max distance" values, etc.<br/>
                                    <br/>
                                    NOTE: The server supports not more than several queries per-minute per-user. If you
                                    need to batch-process many variants, please use the
                                    <a href="https://github.com/broadinstitute/SpliceAI-lookup/blob/master/README.md#local-install" target="_blank">source code</a>
                                    to set up your own local instance of the API server or just run the underlying SpliceAI model directly.<br/>
                                    <br/>
                                    For more information about the SpliceAI model, see <a
                                        href="https://doi.org/10.1016/j.cell.2018.12.015" target="_blank">Jaganathan et
                                    al, Cell 2019</a> and <a href="https://github.com/Illumina/SpliceAI" target="_blank">github.com/Illumina/SpliceAI</a>,
                                    and this <a href="https://www.youtube.com/watch?v=oJvhj-tYbBI" target="_blank">recorded
                                    talk</a> by Kishore Jaganathan.<br/>
                                    <br/>
                                    Since February 2023, <a href="https://github.com/tkzeng/Pangolin" target="_blank">Pangolin scores</a> from [<a
                                        href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02664-4"
                                        target="_blank">Zeng & Li 2022</a>] are also shown below the SpliceAI results.
                                </div>
                            </div>
                            <br/>
                            <div style="padding-left: 0px; color: gray">
                                <table>
                                    <thead style="text-align: left"><tr><th style="padding-bottom: 20px; min-width:430px">Examples &nbsp; (on hg38):</th><th></th></tr></thead>
                                    <tbody>
                                    <tr><td>chr8-140300616-T-G</td><td></td></tr>
                                    <tr><td>chr8 &nbsp; 140300616 &nbsp; T &nbsp; G</td><td></td></tr>
                                    <tr>
                                        <td style="padding-right: 0px">NM_001089.3(ABCA3):c.875A>T (p.Glu292Val)</td>
                                        <td style="padding-left: 0px">
                                            <a id="more-details2-button" href="#"
                                               onclick="$('.more-details2').show();$('#more-details2-button').hide();">[show more examples]</a>
                                        </td>
                                    </tr>
                                    <tr style="display:none;" class="more-details2"><td>8:140300616 T&gt;G</td><td>'chr' <i>prefix is optional</i></td></tr>
                                    <tr style="display:none;" class="more-details2"><td>1:930130:C:G</td><td><i>position with overlapping genes</i></td></tr>
                                    <tr style="display:none;" class="more-details2"><td>1-1042601-A-AGAGAG</td><td><i>insertion of GAGAG</i></td></tr>
                                    <tr style="display:none;" class="more-details2"><td>1-1042466-GGGC-G</td><td><i>deletion of GGC</i></td></tr>
                                    <tr style="display:none;" class="more-details2"><td>NM_000249.4(MLH1):c.116G>A</td><td><i>another HGVS example</i></td></tr>
                                    </tbody>
                                </table>
                            </div>
                        </div>
<!--
                        <br />
                        <div style="font-size: large; color: darkred">
                          <br />
                          <b>The server is currently down for updates</b><br />
                          <br />
                          Please check back in ~1/2 hour.
                        </div>
                        <br />
                        <br />
-->
                        <div class="ui form" style="width:100%">
                            <div class="ui input" style="padding-bottom:30px; padding-right:8px; width:100%">
                                <input id="search-box" type="text" style="width: 100%; padding: 7px 10px" placeholder="Enter a variant...">
                            </div>
                            <div class="ui stackable grid">
                                <div class="row">
                                    <div class="thirteen wide column">
                                        <div class="field" style="padding-right:10px; white-space:nowrap">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Genome version:</b></span>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="hg" value="37"/><label>hg19</label>
                                            </div>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="hg" value="38" checked/><label>hg38</label>
                                            </div>
                                        </div>
                                        <div class="field" style="padding-right:10px; white-space:nowrap">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Score type:</b></span>
                                            <div class="ui radio checkbox"
                                                 style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="mask" value="1" checked/><label>masked</label>
                                            </div>
                                            <div class="ui radio checkbox"
                                                 style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="mask" value="0"/><label>raw</label>
                                            </div>
                                            <i style="margin-left:20px" class="question circle outline icon" data-position="right center"
                                               data-content="Splicing changes corresponding to strengthening annotated splice sites and weakening unannotated splice sites are typically much less pathogenic than weakening annotated splice sites and strengthening unannotated splice sites. Selecting 'masked' will hide the score for such splicing changes and show 0 instead. Selecting 'raw' will show all scores. SpliceAI developers recommend using 'raw' scores for alternative splicing analysis and 'masked' scores for variant interpretation."></i>
                                        </div>
                                        <div class="field" style="padding-right:10px; white-space:nowrap">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Max distance:</b></span>
                                            <div class="ui input"
                                                 style="padding-left:15px; padding-bottom:10px; padding-right:8px; width:100px">
                                                <input id="max-distance-input" type="text" value="500" style="padding:7px 10px">
                                            </div>
                                            <i class="question circle outline icon" data-position="right center"
                                               data-content="For each variant, SpliceAI looks within a window (+/- 500bp by default) to see how the variant affects the probabilities of different positions being splice acceptors or donors. The distance specified here controls the size of this window. The maximum allowed value is 10,000bp."></i>
                                        </div>
                                    </div>
                                    <div class="three wide column">
                                        <button id="submit-button" class="ui primary button" style="margin-bottom: 15px">Submit</button>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        </div>
        <div class="four wide column">
            <div class="only-large-screen" style="padding:25px 0px">
                <div>
                    <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues" target="_blank">SpliceAI-lookup/issues</a>: issues or feature requests for this website.
                    <a href="https://github.com/illumina/SpliceAI" target="_blank">SpliceAI/issues</a>: issues with the underlying SpliceAI model.<br/>
                    <br/>
                    <b>Sep 14, 2023</b><br/>
                    - <a href="https://github.com/pj-sullivan">@pj-sullivan</a> added controls to show/hide transcripts<br/>
                    - added MANE Select transcript labels<br/>
                    - updated to gencode v44 "basic" annotations (instead of comprehensive). These have fewer transcripts per gene.<br/>
                    - added demo of igv.js visualizations. TODO: allow users to select which tracks to show.<br/>
                    <br/>

                    <a id="more-details3-button" href="#"
                       onclick="$('#more-details3').show();$('#more-details3-button').hide();">[show older updates]</a>
                    <div id="more-details3" style="display:none">
                        <b>May 22, 2023</b><br/>
                        - changed defaults to 'masked' and max distance to 500bp<br/>
                        - retired <b>Illumina precomputed scores</b> since they were created using Gencode v24
                        and max distance = 50bp and so can differ from computed scores which are based on the latest Gencode.<br/>
                        <br/>
                        <b>May 12, 2023</b><br/>
                        - added REF & ALT score columns to show SpliceAI's underlying predictions for each haplotype
                        (using <a href="https://github.com/Illumina/SpliceAI/pull/92">SpliceAI PR by @h-joshi</a>).
                        The &#916;&nbsp; score column is the difference between these two scores.<br/>
                        - added support for MNPs (see <a
                            href="https://github.com/broadinstitute/SpliceAI-lookup/issues/1">issue #1</a> & <a
                            href="https://github.com/Illumina/SpliceAI/pull/119">SpliceAI PR by @kdahlo</a>)<br/>
                        - updated to <a href="https://www.gencodegenes.org/human/releases.html">Gencode v43</a> (was previously on v42)<br/>
                        <br/>
                        <b>May 10, 2023</b><br/>
                        - fixed bug where "raw" Pangolin scores were shown regardless of whether the user selected "raw" or "masked".<br/>
                        <br/>

                        <b>Feb 5, 2023</b><br/>
                        - show <a href="https://github.com/tkzeng/Pangolin">Pangolin scores</a> in addition to SpliceAI scores.<br/>
                        <br/>
                        <b>June 2, 2021</b><br/>
                        - show RefSeq ids for the subset of Ensembl ENST ids (~30%) that have one or more matching
                        RefSeq NM ids according to Ensembl. See <a
                            href="https://github.com/broadinstitute/SpliceAI-lookup/issues/8">issue #8</a> for details.<br/>
                        <br/>
                        <b>April 11, 2021</b><br/>
                        - show gray background for non-coding transcripts<br/>
                        - switch to showing all non-coding transcripts. (Previously, transcripts with Gencode biotypes
                        like <i>lncRNA</i>, <i>processed_pseudogene</i>, <i>processed_transcript</i>,
                        <i>retained_intron</i>, and
                        <i>nonsense_mediated_decay</i> were filtered out if they overlapped <i>protein_coding</i> transcripts).
                        See <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues/6#issuecomment-816942824">issue #6</a> for details.<br/>
                    </div>
                </div>
                <br/>
                <div>
                    <b>Related web tools:</b><br/>
                    <a href="https://liftover.broadinstitute.org" target="_blank">liftover</a>: for variants/positions/intervals (hg19 <=> hg38 <=> T2T)<br/>
                    <a href="https://cma-search.broadinstitute.org" target="_blank">CMA search</a>: search OMIM by interval, gene or phenotype<br/>
                    <a href="https://tgg-viewer.broadinstitute.org" target="_blank">TGG Viewer</a>: igv.js-based web viewer for public reference tracks and, optionally, private data in Google Storage buckets<br/>
                </div>
            </div>
        </div>
        <div class="one wide column only-large-screen"></div>
    </div>
    <div class="row">
        <div class="one wide column only-large-screen"></div>
        <div class="fourteen wide column">
            <div class="results-div">
                <div id="response-box">
                    <table id="spliceai-table" class="ui stackable table">
                        <thead id="spliceai-header">
                        <tr>
                            <th style="background-color: white;" colspan="7">SpliceAI scores: <a href="https://www.cell.com/cell/pdf/S0092-8674(18)31629-5.pdf" target="_blank">
                                <i class="question circle outline icon" data-position='right center'
                                   data-content="SpliceAI scores are computed for each transcript on the server by running the SpliceAI model with the user-selected parameters (ie. max distance)."></i></a>
                            </th>
                        </tr>
                        <tr>
                            <th>Variant</th>
                            <th>Gene
                                <div style="float:right;">
                                    <span class='transcript-color-legend'
                                          style="background-color: var(--MAIN_TRANSCRIPT_BACKGROUND_COLOR)!important">&nbsp;</span>
                                    = <span class="main-transcript-label">MANE Select</span> transcript
                                    <span class='transcript-color-legend'
                                          style="background-color: var(--NON_CODING_TRANSCRIPT_BACKGROUND_COLOR)!important">&nbsp;</span>
                                    = non-coding transcript
                                </div>
                            </th>
                            <th>&#916;&nbsp; type</th>
                            <th style="white-space: nowrap">&#916;&nbsp; score<i class='question circle outline icon'
                                                                                 data-content="Delta scores range from 0 to 1 and can be interpreted as the probability that the variant affects splicing at any position within a window around it (+/- 500bp by default). In the SpliceAI paper, a detailed characterization is provided for 0.2 (high recall), 0.5 (recommended), and 0.8 (high precision) cutoffs. Consequently, scores of 0.2 or higher appear in green, 0.5 or higher appear in yellow, and 0.8 or higher appear in red."></i>
                            </th>
                            <th style="white-space: nowrap">pre-mRNA position<i class='question circle outline icon'
                                                                                data-content="For each variant, SpliceAI looks within a window (+/- 500bp by default) to see how the variant affects the probabilities of different positions in the pre-mRNA being splice acceptors or donors. The offsets in this column represent positions with the biggest change in probability within the window. Negative values are to the left of the variant in genomic coords, regardless of transcript strand."></i>
                            </th>
                            <th style="white-space: nowrap">REF score<i class='question circle outline icon'
                                                                        data-content="SpliceAI's computed probability that the given pre-mRNA position is a splice acceptor or donor based on the reference haplotype sequence. SpliceAI computes delta scores by taking the difference between the REF and ALT score at each position."></i>
                            </th>
                            <th style="white-space: nowrap">ALT score<i class='question circle outline icon'
                                                                        data-content="SpliceAI's computed probability that the given pre-mRNA position is a splice acceptor or donor based on the alternate haplotype sequence. SpliceAI computes delta scores by taking the difference between the REF and ALT score at each position."></i>
                            </th>
                        </tr>
                        </thead>
                    </table>
                    <table id="transcript-button-table" class="ui stackable table" style="border: none !important">
                        <tbody>
                        <tr>
                            <td colspan="7" style="padding: 10px 0px 25px 0px">
                                <div class="ui buttons">
                                    <button id="main-transcript-button" class="ui button" onclick="updateTranscriptButtons('main')">
                                        <span class="main-transcript-label">MANE Select</span> Transcript
                                    </button>
                                    <div class="or"></div>
                                    <button id="all-transcript-button" class="ui button" onclick="updateTranscriptButtons('all')">
                                        All Transcripts
                                    </button>
                                </div>
                            </td>
                        </tr>
                        </tbody>
                    </table>
                    <table id="pangolin-table" class="ui stackable table">
                        <thead id="pangolin-header">
                        <tr>
                            <th style="background-color: white;" colspan="7">Pangolin scores: <a
                                    href='https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02664-4'
                                    target="_blank"><i class='question circle outline icon' data-position='right center'
                                                       data-content="Pangolin scores are computed for each transcript on the server by running the SpliceAI model with the user-selected parameters (ie. max distance)."></i></a>
                            </th>
                        </tr>
                        <tr>
                            <th>Variant</th>
                            <th>Gene</th>
                            <th>&#916;&nbsp; type</th>
                            <th style="white-space: nowrap">&#916;&nbsp; score<i class='question circle outline icon'
                                                                                 data-content="Delta scores range from 0 to 1 and can be interpreted as the probability that the variant affects splicing at any position within a window around it (+/- 500bp by default). Pangolin does not recommend specific score thresholds; however, here scores of 0.2 or higher appear in green, 0.5 or higher appear in yellow, and 0.8 or higher appear in red."></i>
                            </th>
                            <th style="white-space: nowrap">pre-mRNA position<i class='question circle outline icon'
                                                                                data-content="For each variant, Pangolin looks within a window (+/- 500bp by default) to see how the variant affects the probabilities of different positions in the pre-mRNA being splice sites. The offsets in this column represent positions with the biggest change in probability within the window. Negative values are upstream (5&#039;) of the variant, and positive are downstream (3&#039;) of the variant."></i>
                            </th>
                            <th style="white-space: nowrap">REF score<i class='question circle outline icon'
                                                                        data-content="Working on it!"></i></th>
                            <th style="white-space: nowrap">ALT score<i class='question circle outline icon'
                                                                        data-content="Working on it!"></i></th>
                        </tr>
                        </thead>
                    </table>
                </div>
                <div id="error-box" style="color:darkred; padding:10px 0px; font-size:medium"></div>
                <div id="igvDiv" style="width:100%; padding-top:25px"></div>
            </div>
        </div>
        <div class="one wide column only-large-screen"></div>
    </div>
</div>

<script>
    const VARIANT_RE = new RegExp(
        "^[\\s]*" +
        "(chr)?([0-9XYMTt]{1,2})" +
        "[-\\p{Pd}\\s:]+" +
        "([0-9,]+)" +
        "[-\\p{Pd}\\s:]*" +
        "([ACGT]+)" +
        "[-\\p{Pd}\\s:>]+" +
        "([ACGT]+)",
        'iu'
    )

    const TRANSCRIPT_PRIORITY = {
        "MS": 3,
        "MP": 2,
        "C": 1,
        "N": 0,
    }

    formatScore = (score) => {
        if (score == null) {
            return ""
        }
        return parseFloat(score).toFixed(2)
    }

    getScoreStyle = (score) => {
        score = parseFloat(score)

        if (Math.abs(score) < 0.01) {
            return `style='color:#BBBBBB;'`
        }

        let color
        if (Math.abs(score) >= 0.8) {
            color = "#fccfb8"
        } else if (Math.abs(score) >= 0.5) {
            color = "#fff19d"
        } else if (Math.abs(score) >= 0.2) {
            color = "#cdffd7"
        } else {
            return ""
        }
        return `style='background-color:${color};'`
    }

    getUCSCBrowserUrl = (genomeVersion, chrom, pos) => {
        const genomeVers = genomeVersion.replace('37', '19')
        const chromWithoutPrefix = chrom.toUpperCase().replace('CHR', '')

        return `https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg${genomeVers}&position=chr${chromWithoutPrefix}:${pos}`
    }

    showError = (errorMessage) => {
        $("#error-box").html( $("#error-box").html() + `<br />${errorMessage.toString()}<br />`)
        $("#error-box").show()
    }

    generateResultsTable = async (normalizedVariant, tool, variant, genomeVersion, mask, maxDistance) => {
        /* Generate the results table to show either the SpliceAI or Pangolin scores */

        const variantTokens = (normalizedVariant || "---").split("-")
        const chrom = variantTokens[0]
        const pos = variantTokens[1]
        const ref = variantTokens[2]
        const alt = variantTokens[3]

        const urlArgs = `hg=${genomeVersion}&distance=${maxDistance}&mask=${mask}&variant=${chrom}-${pos}-${ref}-${alt}&raw=${variant}`

        let apiResponse
        try {
            apiResponse = await fetch(`${baseApiUrl}/${tool.toLowerCase()}/?${urlArgs}`)
        } catch(e) {
            throw Error(`${tool} API call failed: Unable to reach server`)
        }

        const apiResponseJson = await apiResponse.json()
        console.log(`${tool} API response:`, apiResponseJson)

        if (!apiResponse.ok || apiResponseJson.error) {
            throw Error(`${tool} API call error ${apiResponseJson.error ? `: ${apiResponseJson.error}` : ""}`)
        }

        //if (tool == "SpliceAI") throw Error(`${tool} error!!`)  // for testing API errors

        //sort transcripts
        apiResponseJson.scores = _(apiResponseJson.scores).sortBy((s) => (  //compute sort key
                100*(s['t_priority'].startsWith("M") ? 0 : 1) +
                10* (s['t_type'] == "protein_coding" ? 0 : 1) +
                -1*  TRANSCRIPT_PRIORITY[s['t_priority']]
        )).values()


        const TRANSCRIPT_PRIORITY_VIEW_LOOKUP = {
            "MS": `<a href="https://www.ncbi.nlm.nih.gov/refseq/MANE" target="_blank">MANE Select transcript</a>`,
            "MP": `<a href="https://www.ncbi.nlm.nih.gov/refseq/MANE" target="_blank">MANE Plus Clinical transcript</a>`,
            "C": "Canonical transcript",
        }

        const transcriptCategories = {}
        const tableRows = []
        for (const scores of apiResponseJson.scores) {
            const isPangolin = tool.toLowerCase() == "pangolin"
            const subRowCount = isPangolin ? 2 : 4
            const strand = scores['t_strand'] == "-" ? "minus" : "plus"
            const gnomadVersion = genomeVersion === "38" ? "gnomad_r3" : "gnomad_r2_1"
            const isMainTranscript = scores['t_priority'] != "N" && (scores['t_priority'] != "C" || transcriptCategories["MS"] == undefined)
            const refSeqLink = scores['t_refseq_ids']? `/ <a href="https://www.ncbi.nlm.nih.gov/search/all/?term=${scores['t_refseq_ids'][0]}" target="_blank">${scores['t_refseq_ids'][0]}</a>` : ""

            transcriptCategories[scores['t_priority']] = true

            const resultRowClasses = [`${tool.toLowerCase()}-result-row`]
            if (isMainTranscript) {
                resultRowClasses.push("main-transcript")
            } else {
                resultRowClasses.push("non-main-transcript")
            }
            if(scores['t_type'] == "protein_coding") {
                resultRowClasses.push("coding-transcript")
            } else {
                resultRowClasses.push("non-coding-transcript")
            }

            $(`#${tool.toLowerCase()}-header`).nextAll().remove()
            let row = `<tr class="${resultRowClasses.join(' ')}">
                <td rowspan="${subRowCount}" style="vertical-align: top">
                    <div style="margin-right:10px; display: inline-block">${variant}</div><br class="only-large-screen"/>
                    <br class="only-large-screen"/>
                    <a href="${getUCSCBrowserUrl(genomeVersion, chrom, pos)}" class="small-link" target="_blank">UCSC</a>,
                    <a href="https://gnomad.broadinstitute.org/variant/${variant}?dataset=${gnomadVersion}" class="small-link" target="_blank">gnomAD</a>
                </td>
                <td rowspan="${subRowCount}" style="vertical-align: top">
                    <div style="margin-right:10px; display: inline-block">
                        ${scores['g_name']}
                            <div class="small-link"> (
                                <a href="https://useast.ensembl.org/Homo_sapiens/Gene/Summary?g=${scores['g_id'].split('.')[0]}" target="_blank">${scores['g_id']}</a>
                                 / <a href="https://useast.ensembl.org/Homo_sapiens/Transcript/Summary?t=${scores['t_id'].split('.')[0]}" target="_blank">${scores['t_id']}</a>
                                 ${refSeqLink})
                            </div><br />
                            <br class="only-large-screen" />
                        <div class="small-link"><a href="https://www.gencodegenes.org/pages/biotypes.html" target="_blank">${scores['t_type'].replace(/_/g, " ")}</a></div>
                        <div class="small-link">${scores['t_priority'] != "N" ? TRANSCRIPT_PRIORITY_VIEW_LOOKUP[scores['t_priority']] : ""}</div>
                        <div class="small-link"> (${strand} strand)</div>
                    </div><br class="only-large-screen" />
                    <br class="only-large-screen" />
                    <a href="https://www.omim.org/search?search=${scores['g_name']}" class="small-link" target="_blank">OMIM</a>,
                    <a href="https://gtexportal.org/home/gene/${scores['g_name']}" class="small-link" target="_blank">GTEx</a>,
                    <a href="https://gnomad.broadinstitute.org/gene/${scores['g_name']}?dataset=${gnomadVersion}" class="small-link" target="_blank">gnomAD</a>,
                    <a href="https://search.clinicalgenome.org/kb/genes?page=1&size=25&search=${scores['g_name']}" class="small-link" target="_blank">ClinGen</a>,
                    <a href="https://useast.ensembl.org/Homo_sapiens/Gene/Summary?g=${scores['g_name']}" class="small-link" target="_blank">Ensembl</a>,
                    <a href="https://decipher.sanger.ac.uk/gene/${scores['g_name']}" class="small-link" target="_blank">Decipher</a>,
                    <a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=${scores['g_name']}" class="small-link" target="_blank">GeneCards</a>
                </td>`

            for (const [i, label, DS, DP, RAW_REF, RAW_ALT] of (
                isPangolin ? [
                    [0, "Splice Loss", "DS_SL", "DP_SL", "SL_REF", "SL_ALT"],
                    [1, "Splice Gain", "DS_SG", "DP_SG", "SG_REF", "SG_ALT"],
                ] : [
                    [0, "Acceptor Loss", "DS_AL", "DP_AL", "DS_AL_REF", "DS_AL_ALT"],
                    [1, "Donor Loss",    "DS_DL", "DP_DL", "DS_DL_REF", "DS_DL_ALT"],
                    [2, "Acceptor Gain", "DS_AG", "DP_AG", "DS_AG_REF", "DS_AG_ALT"],
                    [3, "Donor Gain",    "DS_DG", "DP_DG", "DS_DG_REF", "DS_DG_ALT"],
                ])) {
                if (i > 0) {
                    row += `</tr><tr class="${resultRowClasses.join(' ')}">`
                }
                row += `
                        <td>${label}</td>
                        <td ${getScoreStyle(scores[DS])}>${formatScore(scores[DS])}</td>
                        <td>${
                            (
                                parseFloat(scores[DS]) == 0 &&
                                parseFloat(scores[RAW_REF]) == 0 &&
                                parseFloat(scores[RAW_ALT]) == 0
                            )? "" : scores[DP] + " bp"}
                        </td>
                        <td>${formatScore(scores[RAW_REF])}</td>
                        <td>${formatScore(scores[RAW_ALT])}</td>
                `
            }
            row += "</tr>"

            tableRows.push(row)
        }

        $("#response-box").show()
        $(`#${tool.toLowerCase()}-header`).after(tableRows.join(""))

        if (transcriptCategories["MS"]) {
            $(".main-transcript-label").html("MANE Select")
        } else if (transcriptCategories["MP"]) {
            $(".main-transcript-label").html("MANE Plus Clinical")
        } else if (transcriptCategories["C"]) {
            $(".main-transcript-label").html("Canonical")
        } else {
            $("#transcript-button-table").hide()
        }
        if (Object.keys(transcriptCategories).length > 1) {
            $("#transcript-button-table").show()
            updateTranscriptButtons("main")
        }

        return apiResponseJson
    }

    updateTranscriptButtons = (category) => {
        $("#main-transcript-button, #all-transcript-button").removeClass("primary")
        $(`#${category}-transcript-button`).addClass("primary")

        if (category == "main") {
            $(".spliceai-result-row").hide()
            $(".spliceai-result-row.main-transcript").show()
        } else if (category == "all") {
            $(".spliceai-result-row").show()
        }
    }

    // define urlParam function for parsing url parameters like ?variant=chr1-1234567-T-G
    $.urlParam = (name) => {
        const results = new RegExp('[\?&#]?' + name + '=([^&#]*)').exec(window.location.hash);
        return (results !== null) ? decodeURIComponent(results[1]) || 0 : false;
    }

    normalizeVariant = async (variant, genomeVersion) => {
        /* Convert the given variant to a standardized "{chrom}-{pos}-{ref}-{alt}" string using a regular expression, or
         * if that fails, assume the variant is in HGVS notation and try using an Ensembl API to convert it.
         *
         * Args:
         *  variant (string): user input text
         *  genomeVersion (string): "37" or "38"
         *
         * Return:
         *  string: the input variant reformatted as a "{chrom}-{pos}-{ref}-{alt}" string
         */

        const matchedRegExp = variant.match(VARIANT_RE)
        if (matchedRegExp) {
            // was able to parse the user input using a simple reg-exp
            const chrom = matchedRegExp[2].toUpperCase()
            const pos = parseInt(matchedRegExp[3].replace(/,/g, ""))
            const ref = matchedRegExp[4]
            const alt = matchedRegExp[5]

            return `${chrom}-${pos}-${ref}-${alt}`
        }

        // otherwise try calling the Ensembl API on the user input
        const ensemblApiUrl = `https://${genomeVersion == '37' ? 'grch37.' : ''}rest.ensembl.org/vep/human/hgvs/${variant.trim()}?content-type=application/json&vcf_string=1`

        let ensemblApiResponse
        try {
            ensemblApiResponse = await fetch(ensemblApiUrl)
        } catch(e) {
            console.error(e)
            throw Error(`Ensembl API call failed: Unable to reach server`)
        }

        const ensemblApiResponseJson = await ensemblApiResponse.json()
        console.log(`Ensembl API response:`, ensemblApiResponseJson)

        if (!ensemblApiResponse.ok || ensemblApiResponseJson.error) {
            let errorText =`${ensemblApiResponseJson.error}`
            const unableToParseHGVSErrorMatch = errorText.match("Unable to parse HGVS notation")
            if (unableToParseHGVSErrorMatch) {
                errorText = `Ensembl API is unable to parse the variant ${variant}`
            }
            const refAlleleErrorMatch = errorText.match(
                new RegExp("[(]([ACGTRYSWKMBDHVN]+)[)] does not match reference allele given by HGVS notation"))
            if (refAlleleErrorMatch) {
                errorText = `${variant} has an unexpected reference allele. The hg${genomeVersion} reference allele should be ${refAlleleErrorMatch[1]}`
            }

            throw new Error(errorText);
        }

        if (!ensemblApiResponseJson[0] || !ensemblApiResponseJson[0].vcf_string) {
            throw new Error(`Unexpected response: ${ensemblApiResponseJson}`);
        }
        return ensemblApiResponseJson[0].vcf_string
    }

    const generateIgvConfig = (spliceaiResponseJson, pangolinResponseJson, genomeVersion) => {
        /* Generates an igv.js config json objects based on the predicted scores from SpliceAI and/or Pangolin
        *
        * Args:
        *   allNonZeroScoresFromSpliceAI (array): an array of acceptor/donor loss/gain scores from SpliceAI
        *   allNonZeroScoresFromPangolin (array): an array of acceptor/donor loss/gain scores from Pangolin
        *   genomeVersion (string): "37" or "38"
        *
        * Return:
        *   object: an igv.js config json object
        */

        const apiResponseJson = spliceaiResponseJson || pangolinResponseJson
        let chrom = apiResponseJson.chrom
        chrom = `chr${chrom.replace('chr', '')}`
        const variantPos = apiResponseJson.pos
        const variantRef = apiResponseJson.ref
        const variantAlt = apiResponseJson.alt
        

        let minPos = 1e9
        let maxPos = 0
        for (const apiResponse of [spliceaiResponseJson, pangolinResponseJson]) {
            if (!apiResponse) {
                continue
            }
            for (const scores of apiResponse.allNonZeroScores) {
                scores["chr"] = chrom
                scores["start"] = scores["pos"]
                scores["end"] = scores["pos"]
                minPos = Math.min(minPos, scores["pos"])
                maxPos = Math.max(maxPos, scores["pos"])
            }
        }

        // specify IGV tracks and the data to display in them
        const tracks = []

        tracks.push({
            name: "Refseq",
            format: "refgene",
            url: genomeVersion == "38" ?
                "https://hgdownload.soe.ucsc.edu/goldenPath/hg38/database/ncbiRefSeq.txt.gz" :
                "https://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/ncbiRefSeq.txt.gz",
            indexed: false,
            infoURL: "https://www.ncbi.nlm.nih.gov/gene/?term=$$",
            height: 100,
        })

        tracks.push({
            name: "Variant (?) ",
            type: "vcf",
            description: `This tracks shows the position of the <b>${chrom}:${variantPos} ${variantRef}>${variantAlt}</b> variant.`,
            height: 30,
            features: [
                {
                    chr: chrom,   //see createVCFVariant function in the igv.js repo for the allowed fields
                    pos: variantPos,
                    start: variantPos - 1,
                    end: variantPos,
                    referenceAllele: variantRef,
                    alternateBases: variantAlt,
                    names: ".", // id in VCF
                    info: {
                        variant: `${chrom}-${variantPos}-${variantRef}-${variantAlt}`,
                    }
                },
            ],
        })

        if (spliceaiResponseJson) {
            tracks.push({
                name: `SpliceAI (?)`,
                description: `<b>Transcript: ${spliceaiResponseJson.allNonZeroScoresTranscriptId}</b> (${spliceaiResponseJson.allNonZeroScoresStrand == '-' ? 'minus' : 'plus'} strand)<br />
                                SpliceAI predictions of splice acceptor & donor sites within the +/- ${spliceaiResponseJson.distance}bp window<br />
                                around <b>${chrom}:${variantPos} ${variantRef}>${variantAlt}</b>. This track shows markers that represent the scores for the<br/>
                                <span style="color:#0000B4"><b>reference sequence</b></span> (without the variant) and the
                                <span style="color:#05d0d2"><b>alternate sequence</b></span> (with the variant). <br />
                                A marker is shown at all positions where a score is ≥ 0.01. <br/>
                                SpliceAI outputs separate predictions for whether a given position is a splice acceptor and/or a splice donor, so<br/>
                                <b>A</b>'s represent acceptor probability scores, and <br/>
                                <b>D</b>'s represent donor probability scores.<br/>
                                The letters are shown up-side-down when the alternate sequence score is lower than the reference sequence score.<br/>
                                Numerical labels show the scores predicted from the reference sequence.
                                `,
                height: 100,
                rawOrDelta: "raw",
                tool: "SpliceAI",
                type: "spliceprediction",
                features: spliceaiResponseJson.allNonZeroScores,
                strand: spliceaiResponseJson.allNonZeroScoresStrand,
            })
            tracks.push({
                name: `SpliceAI (?)`,
                description: `<b>Transcript: ${spliceaiResponseJson.allNonZeroScoresTranscriptId}</b> (${spliceaiResponseJson.allNonZeroScoresStrand == '-' ? 'minus' : 'plus'} strand)<br />
                                SpliceAI delta score predictions for gain and loss of splice acceptors and donors within a <br />
                                +/- ${spliceaiResponseJson.distance}bp window around <b>${chrom}:${variantPos} ${variantRef}>${variantAlt}</b>.<br />
                                This track shows markers at all positions where the |delta score| is ≥ 0.01. <br/>
                                <b>Bar colors:</b> <b style="color:#FF0000">red</b> is for delta scores ≥ 0.8,
                                <b style="color:#FFCB1F">yellow</b> is for delta scores ≥ 0.5,
                                <b style="color:#1fb839">green</b> is for delta scores ≥ 0.2<br />
                                <b>A</b>'s represent acceptor probability scores<br/>
                                <b>D</b>'s represent donor probability scores<br/>
                                `,
                height: 200,
                rawOrDelta: "delta",
                tool: "SpliceAI",
                type: "spliceprediction",
                features: spliceaiResponseJson.allNonZeroScores,
                strand: spliceaiResponseJson.allNonZeroScoresStrand,
            })
        }

        if (pangolinResponseJson) {
            tracks.push({
                name: `Pangolin (?)`,
                description: `<b>Gene: ${pangolinResponseJson.allNonZeroScoresTranscriptId}</b> (${pangolinResponseJson.allNonZeroScoresStrand == '-' ? 'minus' : 'plus'} strand)<br />
                                Pangolin delta score predictions for gain and loss of splice sites at positions<br />
                                within a +/- ${pangolinResponseJson.distance}bp window around <b>${chrom}:${variantPos} ${variantRef}>${variantAlt}</b>.<br />
                                This track shows markers at all positions where the delta score is ≥ 0.1. <br/>
                                <b>Bar colors:</b> <b style="color:#FF0000">red</b> is for delta scores ≥ 0.8,
                                <b style="color:#FFCB1F">yellow</b> is for delta scores ≥ 0.5,
                                <b style="color:#1fb839">green</b> is for delta scores ≥ 0.2<br />
                                `,
                height: 200,
                rawOrDelta: "delta",
                tool: "Pangolin",
                type: "spliceprediction",
                features: pangolinResponseJson.allNonZeroScores,
                strand: pangolinResponseJson.allNonZeroScoresStrand,
            })
        }

        const gencodeTrackPath = genomeVersion == "37" ?
            "gs://tgg-viewer/ref/GRCh37/gencode_v44/knownGene_v44.hg19.sorted.txt.gz" :
            "gs://tgg-viewer/ref/GRCh38/gencode_v44/knownGene_v44.hg38.sorted.txt.gz"

        tracks.push({
            name: "Gencode v44",
            format: 'refgene',
            url: gencodeTrackPath,
            indexURL: `${gencodeTrackPath}.tbi`,
            indexed: true,
            searchable: true,
            height: 350,
            visibilityWindow: -1,
            order: 1000001,
            displayMode: 'EXPANDED',
            color: 'rgb(76,171,225)',
        })


        for (const minScore of [0.5, ]) {   //0.2,  hide the >= 0.2 tracks for now

            for (const splicePredictionType of ["loss", "gain"]) {
                tracks.push(
                    {
                        name: `SpliceAI: A or D ${splicePredictionType} ≥ ${minScore} (?)`,
                        description: `This track visualizes Illumina's precomputed SpliceAI score tables for SNVs and small INDELs.<br/>
                             Each genomic location where an SNV or INDEL variant has a SpliceAI score ≥ ${minScore} is shown as the origin of an arrow.<br />
                             The arrow points to the location where that variant most likely causes acceptor or donor ${splicePredictionType}.
                             Clicking on the arrow will display additional information.`,
                        type: "spliceJunctions",
                        height: 100,
                        url: `gs://tgg-viewer/ref/GRCh38/spliceai/spliceai_scores.raw.snps_and_indels.hg38.filtered.sorted.score_${minScore}.splice_${splicePredictionType}.bed.gz`,
                        indexURL: `gs://tgg-viewer/ref/GRCh38/spliceai/spliceai_scores.raw.snps_and_indels.hg38.filtered.sorted.score_${minScore}.splice_${splicePredictionType}.bed.gz.tbi`,
                    }
                )
            }
        }


        for (const [filenamePrefix, tissue, sampleCount] of [
            ["GTEX_muscle.803_samples", "muscle", 803],
            ["GTEX_blood.755_samples", "blood", 755],
            ["GTEX_fibs.504_samples", "fibroblasts", 504],
            ["GTEX_lymphocytes.174_samples", "lymphocytes", 174],
            ["GTEX_brain_cortex.255_samples", "brain cortex", 255],
            //["GTEX_frontal_cortex.209_samples", "frontal cortex", 803],
        ]) {
            tracks.push({
                name: `GTEx ${tissue}: ${sampleCount} samples (?)`,
                description: `This track shows the combined splice junctions from all ${sampleCount} ${tissue} samples available in GTEx v8.
                                Splice junctions are labeled with the total number of RNA-seq reads that supported the junction, summed across the ${sampleCount} samples.
                                The coverage track shows the total number of RNA-seq reads that overlap each position, summed across all samples and divided by ${sampleCount}.`,
                type: "merged",
                height: 100,
                tracks: [{
                    type: "wig",
                    format: "bigwig",
                    url: `gs://tgg-viewer/ref/GRCh38/gtex_v8/${filenamePrefix}.normalized.bigWig`,
                }, {
                    type: 'spliceJunctions',
                    format: 'bed',
                    minJunctionEndsVisible: 1,
                    colorBy: 'strand',
                    minTotalReads: 3,
                    url: `gs://tgg-viewer/ref/GRCh38/gtex_v8/${filenamePrefix}.junctions.bed.gz`,
                    indexURL: `gs://tgg-viewer/ref/GRCh38/gtex_v8/${filenamePrefix}.junctions.bed.gz.tbi`,
                }],
            })
        }

        let locusMargin = maxPos - minPos < 200 ? 15 : 50

        // igv.js reference genome specs are copied from https://igv.org/genomes/genomes.json
        // These must be specified explicitly rather than just setting the reference to "hg19" or "hg38" so that
        // the RefSeq gene track can be the first track (as described in https://github.com/igvteam/igv.js/issues/1518)
        let reference = genomeVersion == "37" ? {
            "id": "hg19",
            "name": "Human (GRCh37/hg19)",
            "fastaURL": "https://igv-genepattern-org.s3.amazonaws.com/genomes/seq/hg19/hg19.fasta",
            "indexURL": "https://igv-genepattern-org.s3.amazonaws.com/genomes/seq/hg19/hg19.fasta.fai",
            "cytobandURL": "https://igv-genepattern-org.s3.amazonaws.com/genomes/seq/hg19/cytoBand.txt",
            "aliasURL": "https://s3.amazonaws.com/igv.org.genomes/hg19/hg19_alias.tab",
            "chromosomeOrder": "chr1, chr2, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chr10, chr11, chr12, chr13, chr14, chr15, chr16, chr17, chr18, chr19, chr20, chr21, chr22, chrX, chrY",
            "tracks": tracks,
        } : {
            "id": "hg38",
            "name": "Human (GRCh38/hg38)",
            "fastaURL": "https://igv-genepattern-org.s3.amazonaws.com/genomes/seq/hg38/hg38.fa",
            "indexURL": "https://igv-genepattern-org.s3.amazonaws.com/genomes/seq/hg38/hg38.fa.fai",
            "cytobandURL": "https://s3.amazonaws.com/igv.org.genomes/hg38/annotations/cytoBandIdeo.txt.gz",
            "aliasURL": "https://s3.amazonaws.com/igv.org.genomes/hg38/hg38_alias.tab",
            "chromosomeOrder": "chr1, chr2, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chr10, chr11, chr12, chr13, chr14, chr15, chr16, chr17, chr18, chr19, chr20, chr21, chr22, chrX, chrY",
            "tracks": tracks,
        }

        return {
            reference: reference,
            locus: `${chrom}:${minPos - locusMargin}-${maxPos + locusMargin}`,
        }
    }


    initIgvBrowser = async (spliceaiResponseJson, pangolinResponseJson, genomeVersion) => {
        if ((spliceaiResponseJson && spliceaiResponseJson.allNonZeroScores) || (
            pangolinResponseJson && pangolinResponseJson.allNonZeroScores)) {
            const igvConfig = generateIgvConfig(spliceaiResponseJson, pangolinResponseJson, genomeVersion)
            if (!window.igvBrowser) {
                // create igv.js browser object if it hasn't been created yet
                window.igvBrowser = await igv.createBrowser(document.getElementById("igvDiv"), igvConfig)
            } else {
                console.log("Updating igv config to", igvConfig)
                await window.igvBrowser.loadSessionObject(igvConfig)
            }
        }

    }


    handleSubmit = async () => {
        // get user inputs from form elements
        const variant = $("#search-box").val().trim()
        if (!variant) return
        const genomeVersion = $("input[name='hg']:checked").val().trim()
        const mask = $("input[name='mask']:checked").val().trim()
        const maxDistance = $("#max-distance-input").val().trim()

        //start processing submit
        $("#submit-button").addClass(["loading", "disabled"])
        $("#response-box, #transcript-button-table, #error-box").hide()
        $("#error-box").html("")

        // make API calls
        try {
            const normalizedVariant = await normalizeVariant(variant, genomeVersion)

            let [spliceaiResponseJson, pangolinResponseJson] = await Promise.all([
                generateResultsTable(normalizedVariant, "SpliceAI", variant, genomeVersion, mask, maxDistance),
                generateResultsTable(normalizedVariant, "Pangolin", variant, genomeVersion, mask, maxDistance),
            ].map(p => p.catch(e => ({'error': e.message}))))  // if an error is thrown, return {'error': e} instead of the JSON response

            if (spliceaiResponseJson.error) {
                showError(`SpliceAI API error: ${spliceaiResponseJson.error}`)
                spliceaiResponseJson = null
                $("#spliceai-table, #transcript-button-table").hide()
            } else {
                $("#spliceai-table, #transcript-button-table").show()
            }

            if (pangolinResponseJson.error) {
                showError(`Pangolin API error: ${pangolinResponseJson.error}`)
                pangolinResponseJson = null
                $("#pangolin-table").hide()
            } else {
                $("#pangolin-table").show()
            }

            // update url params
            window.location.hash = "#" + $.param({
                variant: variant,
                hg: genomeVersion,
                distance: maxDistance,
                mask: mask,
            })

            // update igv.js browser. Use delay to avoid blocking the event thread for too long during page init
            setTimeout(() => initIgvBrowser(spliceaiResponseJson, pangolinResponseJson, genomeVersion), 10)

        } catch(e) {
            console.error(e)
            showError(e.message)
        }

        //done processing submit
        $("#submit-button").removeClass(["loading", "disabled"])
    }

    applyUrlSettingsToFormElements = () => {
        // get optional settings from the url and update form settings
        const hgFromUrl = $.urlParam('hg')
        if (hgFromUrl) {
            $(`input[name='hg'][value='${hgFromUrl}']`).prop("checked", true)
        }

        const maxDistanceFromUrl = $.urlParam('distance')
        if (maxDistanceFromUrl) {
            $("#max-distance-input").val(maxDistanceFromUrl)
        }

        const maskFromUrl = $.urlParam('mask')
        if (maskFromUrl) {
            $(`input[name='mask'][value='${maskFromUrl}']`).prop("checked", true)
        }

        //update the variant input box last and trigger a search
        const variantFromUrl = $.urlParam('variant')
        if (variantFromUrl) {
            $("#search-box").val(variantFromUrl)
            $("#submit-button").click()
        }
    }

    $(document).ready(() => {
        $("#other-version-of-this-site-link").html(isBetaSite ?
            `or <a href="https://spliceailookup.broadinstitute.org">switch to the production version</a> of this site.` :
            `or <a href="https://broadinstitute.github.io/SpliceAI-lookup-dev/index.html">switch to the BETA version</a> of this site to test out future features.`)

        // init UI elements
        $('.ui.radio.checkbox').checkbox()
        $('.question').popup()
        $("#search-box").focus()
        $("#response-box").hide()

        // init event handlers
        $("#submit-button").click(handleSubmit)

        $("#search-box, #max-distance-input").keydown((event) => {
            if ((event.keyCode || event.which) == 13) {
                $("#submit-button").click()
            }
        })

        applyUrlSettingsToFormElements()
    })

</script>
</body>
</html>
